{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Similarity Demo\n",
    "In this demo, we will answer the question:\n",
    "\n",
    "_How does the primary sequence of TMPRSS2 differ between species that one would encounter in a farm environment?_\n",
    "\n",
    "We will address this question using sequence alignment and analysis tools from the [Biopython](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec81) Python library.\n",
    "\n",
    "## Outline\n",
    "\n",
    "* Using the [Biopython tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec385) as reference\n",
    "* Prerequisites\n",
    "    * Reading on PSSMs: [Rice.edu](https://www.cs.rice.edu/~ogilvie/comp571/2018/09/11/pssm.html)\n",
    "\n",
    "### Part 1: Preparing input sequences\n",
    "\n",
    "* Intro to `Bio.Align`\n",
    "* Learn how to filter sequence records in a multiple sequence alignment by:\n",
    "    * Species name\n",
    "    * Sequence snippets\n",
    "* Find the \n",
    "* Generate consensus sequences for the cat sequence\n",
    "\n",
    "### Part 2: Analyzing aligned sequences\n",
    "\n",
    "* Compare human homolog to the mouse\n",
    "    * Compute a [log odds substitution matrix](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec390)\n",
    "    * What are the log odds of the following polymorphisms?\n",
    "        * Hydrophobic -> hydrophilic and vice versa\n",
    "        * Aromatic -> non-aromatic and vice versa\n",
    "    * Construct a more generalized PSSM for the above categories of \"penalized polymorphisms\"\n",
    "        * For instance, we want to parse `R -> Y` and `S -> I` to `hydrophilic -> hydrophobic`\n",
    "\n",
    "## \"Homework\"\n",
    "\"Homework\" is a recommendedation. If you find yourself more interested in a different analysis, say in a comparison of variants **within** Homo sapiens, feel free to do that analysis instead.\n",
    "\n",
    "* Repeat analysis for each of the other domestic species (dog, horse, chicken, etc.)\n",
    "* Generate a \"generalized PSSM\" for the other types of penalized polymorphisms, such as `acidic -> basic`, `bulky -> small`, `aromatic -> non-aromatic`, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're using [Biopython](http://biopython.org/DIST/docs/tutorial/Tutorial.html) again. If the below import commands fail, you might need to install Biopython from the command line:\n",
    "```bash\n",
    "pip install biopython\n",
    "\n",
    "# or using poetry\n",
    "poetry add biopython\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import AlignInfo, MultipleSeqAlignment\n",
    "from Bio import AlignIO, Alphabet, SeqRecord, Seq, SubsMat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part #1\n",
    "____"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read the alignment records\n",
    "We use the Python function from the Biopython package: `Bio.AlignIO.read` to read the trimmed alignment file. This Python function reads the `*.txt` file in the `'fasta'` format and returns an instance of `Bio.Align.MultipleSeqAlignment` (documentation can be found [here](https://biopython.org/DIST/docs/api/Bio.Align.MultipleSeqAlignment-class.html) and [here](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec81))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<<class 'Bio.Align.MultipleSeqAlignment'> instance (9757 records of length 60918, SingleLetterAlphabet()) at 7fa2faa05b10>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment = AlignIO.read(open('./trimmed_alg.txt'), format='fasta')\n",
    "alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each element of this list-like instance is a sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('------------------------------------------------------...---', SingleLetterAlphabet()), id='7260.FBpp0240705', name='7260.FBpp0240705', description='7260.FBpp0240705', dbxrefs=[])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment[0] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This instance of `Bio.Align.MultipleSeqAlignment` is a lot like a Python list. For instance, you can:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of sequence records:  9757\n",
      "number of sequence records (a different way):  9757\n",
      "ID of the 100th sequence:  9796.ENSECAP00000016722\n"
     ]
    }
   ],
   "source": [
    "# get the number of sequences in this alignment\n",
    "print(\"number of sequence records: \", len(alignment))\n",
    "\n",
    "# iterate over the sequence records in the alignment\n",
    "record_counter = 0\n",
    "for record in alignment:\n",
    "    record_counter += 1\n",
    "print(\"number of sequence records (a different way): \", record_counter)\n",
    "\n",
    "# get the 100th sequence record in the alignment\n",
    "print(\"ID of the 100th sequence: \", alignment[99].id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter the sequences in the alignment\n",
    "For now, we're only interested in \"domestic species,\" or species whose scientific name is in the Python list `domestic_sp_names`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "domestic_sp_names = [\n",
    "    'Homo sapiens', # human\n",
    "    'Mus musculus', # mouse\n",
    "    'Canis lupus familiaris', # dog\n",
    "    'Felis catus', # cat\n",
    "    'Bos taurus', # cattle\n",
    "    'Equus caballus', # horse\n",
    "    'Gallus gallus' # chicken\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequences in the `Bio.Align.MultipleSeqAlignment` are for **all** the species that EggNOG could find, including worms, polar bears, and other species that we're not interested in.\n",
    "\n",
    "Let's filter out sequences from species whose names are **not** in the list `domestic_sp_names`. To do this, we will:\n",
    "1. Get the scientific name for each species, and load it into the `description` attribute of each sequence. This should be familiar from the [descriptive stats demo](../descriptive_stats_demo/eggNOG_alignment_metadata.ipynb).\n",
    "2. Use a [list comprehension](https://github.com/wilfredinni/python-cheatsheet#list-comprehension) to get the list of sequences for species that we are interested in.\n",
    "3. The above step will generate a Python list; it will need to be converted to an instance of  `Bio.Align.MultipleSeqAlignment` if we want to use fancy Biopython analysis tools on it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step #1: Get scientific name for each species\n",
    "This should be familiar from the [descriptive stats demo](../descriptive_stats_demo/eggNOG_alignment_metadata.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 20200706_seq_sim.ipynb   extended_members.txt\t tree.txt\n",
      "\"Ty's Playbook.ipynb\"\t  raw_alg.txt\t\t trimmed_alg.txt\n",
      " 20200706_seq_sim.ipynb   extended_members.txt\t tree.txt\n",
      "\"Ty's Playbook.ipynb\"\t  raw_alg.txt\t\t trimmed_alg.txt\n"
     ]
    }
   ],
   "source": [
    "!ls\n",
    "!\n",
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_1</th>\n",
       "      <th>id_2</th>\n",
       "      <th>species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CRE24749</td>\n",
       "      <td>CRE24749</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:DS268562,E3N945_CAERE,E3N945,CRE_24749...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRE21132</td>\n",
       "      <td>CRE-TRY-4</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:E3MEX0,DS268440,E3MEX0_CAERE,CRE21132,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRE24758</td>\n",
       "      <td>CRE-TRY-6</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:E3N963,DS268562,E3N963_CAERE,CRE24758,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRE18672</td>\n",
       "      <td>CRE18672</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:DS268410,E3LKX4_CAERE,E3LKX4,CRE_18672...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRE24729</td>\n",
       "      <td>CRE-TRY-3</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:E3N418,DS268522,E3N418_CAERE,CRE24729,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id_1       id_2                 species         \\\n",
       "0  CRE24749   CRE24749  Caenorhabditis remanei  31234   \n",
       "1  CRE21132  CRE-TRY-4  Caenorhabditis remanei  31234   \n",
       "2  CRE24758  CRE-TRY-6  Caenorhabditis remanei  31234   \n",
       "3  CRE18672   CRE18672  Caenorhabditis remanei  31234   \n",
       "4  CRE24729  CRE-TRY-3  Caenorhabditis remanei  31234   \n",
       "\n",
       "                                                      \n",
       "0  aliases:DS268562,E3N945_CAERE,E3N945,CRE_24749...  \n",
       "1  aliases:E3MEX0,DS268440,E3MEX0_CAERE,CRE21132,...  \n",
       "2  aliases:E3N963,DS268562,E3N963_CAERE,CRE24758,...  \n",
       "3  aliases:DS268410,E3LKX4_CAERE,E3LKX4,CRE_18672...  \n",
       "4  aliases:E3N418,DS268522,E3N418_CAERE,CRE24729,...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmprss2_ext = pd.read_table('../seq_sim_demo/extended_members.txt', header=None)\n",
    "tmprss2_ext.columns = ['id_1', 'id_2', 'species', '', '']\n",
    "tmprss2_ext.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in alignment:\n",
    "    \n",
    "    # while we're at it, let's make sure that Biopython knows these\n",
    "    # are protein sequences\n",
    "    record.seq.alphabet = Alphabet.generic_protein\n",
    "    \n",
    "    # from visual inspection we know the name format is XXXX.unique_id,\n",
    "    # so we split on \".\" and take the last element of the list\n",
    "    id_code = record.id.split('.')[-1]\n",
    "    \n",
    "    # reference the metadata to get the species name\n",
    "    sp_name = tmprss2_ext[tmprss2_ext['id_1'] == id_code]['species'].values\n",
    "    \n",
    "    try:\n",
    "        sp_name = sp_name.item()\n",
    "    except ValueError:\n",
    "        sp_name = None\n",
    "    \n",
    "    # assign the species name to the species attribute\n",
    "    record.description = sp_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step #2: Use a list comprehension to filter to domestic species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "dom_aln_list = [record for record in alignment\n",
    "                if record.description in domestic_sp_names]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that the length of this filtered list is much shorter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of records for all species: 9757\n",
      "number of records for domestic species: 732\n"
     ]
    }
   ],
   "source": [
    "print(\"number of records for all species:\", len(alignment))\n",
    "print(\"number of records for domestic species:\", len(dom_aln_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step #3: Convert this list to a new `MultipleSeqAlignment` instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "dom_aln = MultipleSeqAlignment(dom_aln_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`dom_aln` has the same data, but is a different type of Python variable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dom_aln_list is type: <class 'list'>\n",
      "dom_aln is type: <class 'Bio.Align.MultipleSeqAlignment'>\n"
     ]
    }
   ],
   "source": [
    "print(\"dom_aln_list is type:\", type(dom_aln_list))\n",
    "print(\"dom_aln is type:\", type(dom_aln))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the sequence of human TMPRSS2\n",
    "Before we start comparing sequences to each other, let's get the sequence of TMPRSS2 in `Homo sapiens`. This is the sequence that we will compare other species' homologs to.\n",
    "\n",
    "To do this filtering, let's use a list comprehension, then convert to a `MultipleSeqAlignment`, just like we did before:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_aln_list = [\n",
    "    record for record in dom_aln\n",
    "    if record.description == 'Homo sapiens'\n",
    "]\n",
    "human_aln = MultipleSeqAlignment(human_aln_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that there are many records in the alignment that have `Homo sapiens` as the species:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "118"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(human_aln)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It would be interesting to look at how the differences between these 118 variants _within_ the human species, but let's move on to our inter-species analysis for this demo.\n",
    "\n",
    "## Get the sequence of human isoform 2\n",
    "\n",
    "Let's find the sequence record that has the same sequence as isoform 2 on the [TMPRSS2 UniProt page](https://www.uniprot.org/uniprot/O15393#O15393-1). The first few residues of this isoform are `MPPAPPGG`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "isoform_aln_list = [\n",
    "    record for record in human_aln\n",
    "    if 'MPPAPPGG' in str(record.seq).replace(\"-\", \"\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of human sequences that contain MPPAPPGG: 1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('------------------------------------------------------...---', ProteinAlphabet()), id='9606.ENSP00000381588', name='9606.ENSP00000381588', description='Homo sapiens', dbxrefs=[])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"number of human sequences that contain MPPAPPGG:\", len(isoform_aln_list))\n",
    "human_iso2 = isoform_aln_list[0]\n",
    "human_iso2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is an aligned sequence, so it has a lot of `-` characters that signify residues that are missing relative to other sequences in `alignment`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------MPPAPP----------GGESG-CEE-----------------------------------------------------------R--G-A-A---------GHIEHSRYLS-L-LD---------AV-D----------N----SK-------------------------------------------------------M------------------------------------------------------------------------ALNSG-------------------------S----------P----------------------------------------------------------P---AI---G----P-Y---YENHG----------------------------------------YQPE---------NPY-----------------------------------------------------------------------------------------------------------------------------------------P-A-Q-----------------PT----------------VV-P------------------------------------------------------T----------------------------------V-------------------------------------YE-V-H---P----------A------------------------------QY--Y----P-S--------------P-V-P------Q----YAPRVLT---Q-A--SN--P-------V----V--CTQ--PK-SP---SG-----T-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------V-C--T-----S--KTK--KALCIT----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------L--TLGTFLV------------------------------GAALAAGLLWKFM-GSKCSNSG----------IE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------CD----S-----------------------------------------------------------S------------G--T------C----IN--P-SN-WC--DG----------------------------------------------------------------------------------------------V--------------SH---C-------------------PGG----E----------------DE-N---R--------C-----------V------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------RL-Y-G--P--------------------------------------------------------------------------------------NF---ILQVY-----------------------------------------------------------------------------S----------S---------------------------------------------------------------------------------------------------QRKSWHP--V--CQD------------------D------------------------------------------------------------------WN--------------ENYGRAACRD-----------------------------------------MG-------------------Y----K-NNFYS------------SQGIVDD-----------------------------S--------GSTS-FMKLNTSAGNV------------D----IYK-------KLYH----SDACSSKAVV------------------S-----L---R-CI---------AC----GVN-LNS---------------SRQ---------------------------------------------------------------------------------------------------------------------------------------S--------------------------------R---------I-V---G-GES-A-------L--PG--------A---------------------------------------------------------------WP----WQV---S-LH--------V----------Q---N----------------------------------------------------------------------------V-------------H-------------VCGGS-----------------IIT------P--EW-I-----VT-AAH---CVEK-PL------------------------------NNPW--H-----------------------------------------------------------------------------W-T--A--F-AG-IL-------R-------Q---S--FM-----------------F------------Y----------------------------G-------------------------------------A------GYQ---VEK--VI--------S-H-P---N--------Y-D----------------------S-----KT-K---------N-------------------------------N--------------D--I---A--LM-----------KL--Q---K------P-------------L----------------------------------------------------------------------------------------------T-----F-N--D--L------------V---------------------------K-PV-C-----------------------LPN-----------------------------------------------------------------P--G--------------------M-M--L--Q-P---------E-----QL-C----W------------------------------ISGWG--AT-E-E-------K-G------------------K-------------------------------------------------------------T------------------------------------------------------------------S-E-VLN--------------------------A--A---K----V-LL-IE-TQ-R--------------CN---S----------R--------------------------------Y-VYD------------------N-L-------------------IT-P--A-MI---C-A----------------G-F-----L-Q--G--------N----------------V--D-------SC--------------Q----------GD-----------------------SG-GPLVTSK-----------N--N---------------------------------------I---W------WL---------IGDT-SW-GS------------G-----C-A-K------------A----Y----R----------P-------G-VY---------------------------GN--VMV---F------------T---DWIYRQ---MR--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ADG------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(human_iso2.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can remove these characters using Python's string replacement method, allowing us to more easily look at the amino acid sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MPPAPPGGESGCEERGAAGHIEHSRYLSLLDAVDNSKMALNSGSPPAIGPYYENHGYQPENPYPAQPTVVPTVYEVHPAQYYPSPVPQYAPRVLTQASNPVVCTQPKSPSGTVCTSKTKKALCITLTLGTFLVGAALAAGLLWKFMGSKCSNSGIECDSSGTCINPSNWCDGVSHCPGGEDENRCVRLYGPNFILQVYSSQRKSWHPVCQDDWNENYGRAACRDMGYKNNFYSSQGIVDDSGSTSFMKLNTSAGNVDIYKKLYHSDACSSKAVVSLRCIACGVNLNSSRQSRIVGGESALPGAWPWQVSLHVQNVHVCGGSIITPEWIVTAAHCVEKPLNNPWHWTAFAGILRQSFMFYGAGYQVEKVISHPNYDSKTKNNDIALMKLQKPLTFNDLVKPVCLPNPGMMLQPEQLCWISGWGATEEKGKTSEVLNAAKVLLIETQRCNSRYVYDNLITPAMICAGFLQGNVDSCQGDSGGPLVTSKNNIWWLIGDTSWGSGCAKAYRPGVYGNVMVFTDWIYRQMRADG'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(human_iso2.seq).replace('-', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also notice that most of the sequence of interest is in the middle of the aligned sequence. Let's trim the aligned sequence to generate a compact aligned sequence that it starts with `MPPAPP` and ends with `ADG`. To do this, we will make use of the [`str.index`](https://docs.python.org/2/library/stdtypes.html?highlight=index#str.index) method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index of N-terminus: 33713\n",
      "index of C-terminus: 38856\n"
     ]
    }
   ],
   "source": [
    "index_nterm = str(human_iso2.seq).index('MPPAPP')\n",
    "index_cterm = str(human_iso2.seq).index('ADG')\n",
    "\n",
    "# since we want to cut at ADG^, not ^ADG, we add 3 characters to this index\n",
    "index_cterm += 3\n",
    "\n",
    "print(\"index of N-terminus:\", index_nterm)\n",
    "print(\"index of C-terminus:\", index_cterm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use these indices to trim to the compact sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MPPAPP----------GGESG-CEE-----------------------------------------------------------R--G-A-A---------GHIEHSRYLS-L-LD---------AV-D----------N----SK-------------------------------------------------------M------------------------------------------------------------------------ALNSG-------------------------S----------P----------------------------------------------------------P---AI---G----P-Y---YENHG----------------------------------------YQPE---------NPY-----------------------------------------------------------------------------------------------------------------------------------------P-A-Q-----------------PT----------------VV-P------------------------------------------------------T----------------------------------V-------------------------------------YE-V-H---P----------A------------------------------QY--Y----P-S--------------P-V-P------Q----YAPRVLT---Q-A--SN--P-------V----V--CTQ--PK-SP---SG-----T-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------V-C--T-----S--KTK--KALCIT----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------L--TLGTFLV------------------------------GAALAAGLLWKFM-GSKCSNSG----------IE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------CD----S-----------------------------------------------------------S------------G--T------C----IN--P-SN-WC--DG----------------------------------------------------------------------------------------------V--------------SH---C-------------------PGG----E----------------DE-N---R--------C-----------V------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------RL-Y-G--P--------------------------------------------------------------------------------------NF---ILQVY-----------------------------------------------------------------------------S----------S---------------------------------------------------------------------------------------------------QRKSWHP--V--CQD------------------D------------------------------------------------------------------WN--------------ENYGRAACRD-----------------------------------------MG-------------------Y----K-NNFYS------------SQGIVDD-----------------------------S--------GSTS-FMKLNTSAGNV------------D----IYK-------KLYH----SDACSSKAVV------------------S-----L---R-CI---------AC----GVN-LNS---------------SRQ---------------------------------------------------------------------------------------------------------------------------------------S--------------------------------R---------I-V---G-GES-A-------L--PG--------A---------------------------------------------------------------WP----WQV---S-LH--------V----------Q---N----------------------------------------------------------------------------V-------------H-------------VCGGS-----------------IIT------P--EW-I-----VT-AAH---CVEK-PL------------------------------NNPW--H-----------------------------------------------------------------------------W-T--A--F-AG-IL-------R-------Q---S--FM-----------------F------------Y----------------------------G-------------------------------------A------GYQ---VEK--VI--------S-H-P---N--------Y-D----------------------S-----KT-K---------N-------------------------------N--------------D--I---A--LM-----------KL--Q---K------P-------------L----------------------------------------------------------------------------------------------T-----F-N--D--L------------V---------------------------K-PV-C-----------------------LPN-----------------------------------------------------------------P--G--------------------M-M--L--Q-P---------E-----QL-C----W------------------------------ISGWG--AT-E-E-------K-G------------------K-------------------------------------------------------------T------------------------------------------------------------------S-E-VLN--------------------------A--A---K----V-LL-IE-TQ-R--------------CN---S----------R--------------------------------Y-VYD------------------N-L-------------------IT-P--A-MI---C-A----------------G-F-----L-Q--G--------N----------------V--D-------SC--------------Q----------GD-----------------------SG-GPLVTSK-----------N--N---------------------------------------I---W------WL---------IGDT-SW-GS------------G-----C-A-K------------A----Y----R----------P-------G-VY---------------------------GN--VMV---F------------T---DWIYRQ---MR--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ADG'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_compact = human_iso2[index_nterm:index_cterm]\n",
    "str(human_compact.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These N-terminus and C-terminus indices will be useful when we want to trim sequence records for other species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate consensus sequences for cat homolog\n",
    "\n",
    "Just like the sequence records for `Homo sapiens`, the records for the other `domestic_sp_names` have duplicates. For example, let's look at `Felis catus`, or domesticated cat:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat_aln_list = [\n",
    "    record for record in dom_aln\n",
    "    if record.description == 'Felis catus'\n",
    "]\n",
    "cat_aln = MultipleSeqAlignment(cat_aln_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cat_aln)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's compare 1 sequence, instead of all 100 variants, of cat homolog to the human homolog. To do this, we will generate a **consensus sequence** ([Wikipedia](https://en.wikipedia.org/wiki/Consensus_sequence#:~:text=In%20molecular%20biology%20and%20bioinformatics,position%20in%20a%20sequence%20alignment.)) for the cat variants. We do this in 2 steps:\n",
    "1. Generate a `Bio.Align.AlignInfo.SummaryInfo` instance from the `MultipleSeqAlignment`\n",
    "2. Call the `SummaryInfo` method `dumb_consensus`, which runs a very simple consensus sequence finding algorithm."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step #1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Bio.Align.AlignInfo.SummaryInfo at 0x7fba01ab7580>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_aln_summary = AlignInfo.SummaryInfo(cat_aln)\n",
    "cat_aln_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step #2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...XXX', ProteinAlphabet())"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_aln_consensus = cat_aln_summary.dumb_consensus()\n",
    "cat_aln_consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's use the N-terminus and C-terminus locations that we calculated above to compact this consensus sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'LSRMP-GL-----C-A------------------SVALTL-----P---------------------------------RG------G--------QEGIT-VPSLS-------P--------------MPLQ--------------------------------------------------------------K-------------A----------------QSIP-NGEEL----V--G-------GRVEVFI-----HCPIPHH-------C--------NP---F------H-M-----------------------L------------------TP-----------------------------T---AA---H----C--NAS---------P---A--L-----------------------G---------------I--V---L--------------------EC-----------PTD-----C-----------------------V------G--ARA------PN----------A----S---------------------------------------------------------MP-----Y-------------------------------YC-------------H-P------LH---------------------WHD----T----RLPG-S---I----N--R---------------H-------------D---L---------------------L--G-----------P--------------------------------------------------------------------L-----------------------P----E--G--------------------------------------------------------------------------LP---------------------P--P--------PYP--------Y----Q-----------------RFP------------L-----------R--FD-E--DS-----CP------------GWG-----------------------G------W----P-----------------------------------------------GP-CG--------P-TKG-------GP--I-T-S----I-----------F--D-S----------------------------------------------------------------------------------------------------------------QPSPGN-------L--L-EC-------------------------------------------R--AYFGEGKGPIHV------G-----------------SQD--------L--I--AD--------P-EV--L---KLNTAPN-GS------------TS-----------VS--KRIYFQ------------------C-------Y--------KHQ----------------K-------------------------SPAR------G-------G-P---L-L-------RP------C---G-----------------P---------------RR-A------G----Q-G------------------------------------------------------------IFSC------------S-----S--G---------------------SG--W--P-----C--L----------C--------------------P---ERGH---------Q----------------------RS----SS-------------------------------C-----VV------------G------------------------------------------------------------------L---D--W--------E--------SIPS--G-V---------SW--------------------------------------S-------SG------------C------------------------A------------------------P------------------------------------------------------------------------------T----Q----SYL------Y---------------------W--I------------------L------ET-----------S-T----------------------------------------EVDVEPSLRGPEIVTMGENDPPAAEAPFSFRSLFGLDDLKI--V---------------N-----V---------------------------------------------------------------------------------------AP-QLQD-GSDQPLN-------LF-C--------EVD-------------L-P-------------I----I----DK--F---FSL------YY--SF------------------VCD------------------C-------PD-SPV-TVR-S-----LQV--PA---W---C-D---------F--------------PLGIL-------------KPS------------------M----RR----D----------S---------------------------------------L--I-----I-------------------------AG----------------------------I----CED--G--T----------------------------------P----------------------Q-----G-Y-C---F----W---P------------GHST----P--------------------------------------------------V-C----------------------------------R---------I-----G-G------------ES-G---------------------------------------------------------------------G---P----WQ------L----------------Y-W------HQ---------------W----NPFVPLNAL-------LPL--TAS---------------------SHGD----G--------------H--------------CG--------------------------------W----------AAH---C-------DP-----------------------E-K------------GF---------------------------------------------EY---LQDSD---LIRPS----HYQI-----------G-----------TCHPAPR------------HAGNH--Y-----P-------------------------------------------G----------------------------------------------V---------D-SSW-----H--FNT------N----------------Y-------DP--------------DP-S---------------D--S------------------------------D--I---A--L-------------L--------S-----------EE-CA-------------------RKECVL---------------------------------------------------------------------T-----------GT-------P--------------------------------------P--C-----------------------LP--------------LAI-----------L-----RP--------E------------D-HHL---------------------C-----------------------------------C-------------------------------------GWG------------------------Q--L------T-------------------------------------------------------------EA--------------------------GR----------S--------------------T------------LQ----------------------------------------------------------------C--V-------------------S----------------------------Y---L-E------HE--------L------P------A------------M-VQSC-A-G----HEH-T-----G--------E---G---------K------VGG----E----D--------C-------------------------GD-----------------G-D---SG-GPL-C-----------IS-----VE---------------------------------A------W-------L----------G---SW-G-------------------CQ--------------K--------PA-NNMQEDNQGSP-------G-VY---------------------------T---V-----------------------WI------------------------INEKE-----ARNLQ-PATRNN-P--------SSRWRPSC----PPKTQ----C-----L----TPP-----GFFSTPR-YPLD--Y-----------RI-------EA--ETGSCL-------------------AN-------MSK-----------------------------------------------PAGC------------------'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_consensus_compact = cat_aln_consensus[index_nterm:index_cterm]\n",
    "str(cat_consensus_compact).replace('X', '-')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, this consensus sequence is a `Seq`, not a `SeqRecord`. Let's convert it to a `SeqRecord` so we can compare it to the human sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('LSRMP-GL-----C-A------------------SVALTL-----P--------...---'), id='<unknown id>', name='dumb_consensus', description='Felis catus', dbxrefs=[])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert 'X' to '-' for consistency with human sequence\n",
    "# and convert to a Seq.Seq instance\n",
    "cat_replaced_str =  str(cat_consensus_compact).replace('X', '-')\n",
    "cat_consensus_replaced = Seq.Seq(cat_replaced_str)\n",
    "\n",
    "# then convert to a SeqRecord.SeqRecord instance\n",
    "cat_record_compact = SeqRecord.SeqRecord(cat_consensus_replaced, description='Felis catus', name='dumb_consensus')\n",
    "cat_record_compact"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2: the fun stuff\n",
    "**Finally**, we have human TMPRSS2 and a consensus sequence for cat TMPRSS2. The sequences are aligned and ready for some more advanced analysis with the help of Biopython.\n",
    "\n",
    "Let's start looking at ways we can compare the two sequences. To start, we will answer the question:\n",
    "\n",
    "**At every location in the sequence, what is the percent probability that this position will be a tyrosine (Y), leucine (L), or any other amino acid?**\n",
    "\n",
    "To do this, we will calculate a [**position specific score matrix**](https://www.cs.rice.edu/~ogilvie/comp571/2018/09/11/pssm.html) (PSSM). Let's generate a new, very short `MultipleSeqAlignment` between our human and cat sequences:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<<class 'Bio.Align.MultipleSeqAlignment'> instance (2 records of length 5143, Alphabet()) at 7fba01ab78e0>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hum_cat_aln = MultipleSeqAlignment([human_compact, cat_record_compact])\n",
    "hum_cat_aln"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can generate a `SummaryInfo` instance like we did before, and calculate the PSSM:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Bio.Align.AlignInfo.SummaryInfo at 0x7fa2dfc89dd0>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hum_cat_summary = AlignInfo.SummaryInfo(hum_cat_aln)\n",
    "hum_cat_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Bio.Align.AlignInfo.PSSM at 0x7fa2c07fab50>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hum_cat_pssm = hum_cat_summary.pos_specific_score_matrix(human_compact)\n",
    "hum_cat_pssm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look at the data in the PSSM by inspecting the `pssm` attribute.\n",
    "\n",
    "The PSSM is a Python list, where each element is a [tuple](https://github.com/wilfredinni/python-cheatsheet#tuple-data-type) of length 2. The first element of the tuple is the amino acid in the human sequence, and the second element is a Python [dictionary](https://github.com/wilfredinni/python-cheatsheet#dictionaries-and-structuring-data). The dictionary keys are all the naturally occurring amino acids, and the values are the number of times that amino acid was found at that position in the alignment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## At which positions are the sequences identical?\n",
    "To answer this question, we will use a familiar [for loop](https://github.com/wilfredinni/python-cheatsheet#for-loops-and-the-range-function). When we encounter an `-` in the first element of the `position` tuple, this means that the human sequence had a `-` character at that position. `-` is not an amino acid, so we skip these positions and move on using the [continue statement](https://github.com/wilfredinni/python-cheatsheet#continue-statements).\n",
    "\n",
    "In the `print` statement at the end of the cell, we also make use of [formatted strings](https://github.com/wilfredinni/python-cheatsheet#formatted-string-literals-or-f-strings-python-36) in Python 3.6."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat and human are the same at position 5, which is amino acid P\n",
      "cat and human are the same at position 16, which is amino acid G\n",
      "cat and human are the same at position 24, which is amino acid S\n",
      "cat and human are the same at position 47, which is amino acid A\n",
      "cat and human are the same at position 106, which is amino acid P\n",
      "cat and human are the same at position 157, which is amino acid C\n",
      "cat and human are the same at position 170, which is amino acid C\n",
      "cat and human are the same at position 172, which is amino acid G\n",
      "cat and human are the same at position 176, which is amino acid C\n",
      "cat and human are the same at position 179, which is amino acid G\n",
      "cat and human are the same at position 222, which is amino acid C\n",
      "cat and human are the same at position 281, which is amino acid C\n",
      "cat and human are the same at position 282, which is amino acid G\n",
      "cat and human are the same at position 292, which is amino acid R\n",
      "cat and human are the same at position 293, which is amino acid I\n",
      "cat and human are the same at position 295, which is amino acid G\n",
      "cat and human are the same at position 296, which is amino acid G\n",
      "cat and human are the same at position 302, which is amino acid G\n",
      "cat and human are the same at position 305, which is amino acid P\n",
      "cat and human are the same at position 306, which is amino acid W\n",
      "cat and human are the same at position 307, which is amino acid Q\n",
      "cat and human are the same at position 310, which is amino acid L\n",
      "cat and human are the same at position 316, which is amino acid H\n",
      "cat and human are the same at position 318, which is amino acid C\n",
      "cat and human are the same at position 319, which is amino acid G\n",
      "cat and human are the same at position 327, which is amino acid W\n",
      "cat and human are the same at position 331, which is amino acid A\n",
      "cat and human are the same at position 332, which is amino acid A\n",
      "cat and human are the same at position 333, which is amino acid H\n",
      "cat and human are the same at position 334, which is amino acid C\n",
      "cat and human are the same at position 350, which is amino acid G\n",
      "cat and human are the same at position 371, which is amino acid H\n",
      "cat and human are the same at position 382, which is amino acid D\n",
      "cat and human are the same at position 383, which is amino acid I\n",
      "cat and human are the same at position 384, which is amino acid A\n",
      "cat and human are the same at position 385, which is amino acid L\n",
      "cat and human are the same at position 388, which is amino acid L\n",
      "cat and human are the same at position 400, which is amino acid P\n",
      "cat and human are the same at position 402, which is amino acid C\n",
      "cat and human are the same at position 403, which is amino acid L\n",
      "cat and human are the same at position 404, which is amino acid P\n",
      "cat and human are the same at position 416, which is amino acid C\n",
      "cat and human are the same at position 420, which is amino acid G\n",
      "cat and human are the same at position 421, which is amino acid W\n",
      "cat and human are the same at position 422, which is amino acid G\n",
      "cat and human are the same at position 434, which is amino acid L\n",
      "cat and human are the same at position 447, which is amino acid C\n",
      "cat and human are the same at position 453, which is amino acid Y\n",
      "cat and human are the same at position 461, which is amino acid M\n",
      "cat and human are the same at position 463, which is amino acid C\n",
      "cat and human are the same at position 464, which is amino acid A\n",
      "cat and human are the same at position 465, which is amino acid G\n",
      "cat and human are the same at position 469, which is amino acid G\n",
      "cat and human are the same at position 472, which is amino acid D\n",
      "cat and human are the same at position 474, which is amino acid C\n",
      "cat and human are the same at position 476, which is amino acid G\n",
      "cat and human are the same at position 477, which is amino acid D\n",
      "cat and human are the same at position 478, which is amino acid S\n",
      "cat and human are the same at position 479, which is amino acid G\n",
      "cat and human are the same at position 480, which is amino acid G\n",
      "cat and human are the same at position 481, which is amino acid P\n",
      "cat and human are the same at position 482, which is amino acid L\n",
      "cat and human are the same at position 490, which is amino acid W\n",
      "cat and human are the same at position 492, which is amino acid L\n",
      "cat and human are the same at position 494, which is amino acid G\n",
      "cat and human are the same at position 497, which is amino acid S\n",
      "cat and human are the same at position 498, which is amino acid W\n",
      "cat and human are the same at position 499, which is amino acid G\n",
      "cat and human are the same at position 502, which is amino acid C\n",
      "cat and human are the same at position 508, which is amino acid P\n",
      "cat and human are the same at position 509, which is amino acid G\n",
      "cat and human are the same at position 510, which is amino acid V\n",
      "cat and human are the same at position 511, which is amino acid Y\n",
      "cat and human are the same at position 514, which is amino acid V\n",
      "cat and human are the same at position 520, which is amino acid W\n",
      "cat and human are the same at position 521, which is amino acid I\n"
     ]
    }
   ],
   "source": [
    "# we want to keep track of which amino acid our\n",
    "# \"cursor\" is on in the for loop\n",
    "position_counter = 0\n",
    "\n",
    "for position in hum_cat_pssm.pssm:\n",
    "    \n",
    "    # `position` is the 2-element tuple\n",
    "    # let's give each element a useful name\n",
    "    resi_in_human = position[0]\n",
    "    resi_dict = position[1]\n",
    "    \n",
    "    # skip this position if it is a '-'\n",
    "    # in the human sequence record\n",
    "    if resi_in_human == '-':\n",
    "        continue\n",
    "    else:\n",
    "        # increment the counter by 1\n",
    "        position_counter += 1 \n",
    "    \n",
    "    # if more than one instance of amino acid\n",
    "    # `resi_in_human` was found at this position,\n",
    "    # meaning that the cat homolog is the same amino acid\n",
    "    if resi_dict[resi_in_human] == 2:\n",
    "        print(f\"cat and human are the same at position \" +\n",
    "              f\"{position_counter}, which is amino acid {resi_in_human}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To make sure our `position_counter` variable is working properly, let's double check that the length of the human sequence (without `-` characters) is indeed 529:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the human sequence is 529 amino acids long\n",
      "the human sequence is 529 amino acids long\n"
     ]
    }
   ],
   "source": [
    "# position counter from the above for loop\n",
    "print(f\"the human sequence is {position_counter} amino acids long\")\n",
    "\n",
    "# calling len(str)\n",
    "length_a_different_way = len(str(hum_cat_aln[0].seq).replace('-', ''))\n",
    "print(f\"the human sequence is {length_a_different_way} amino acids long\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that `position_counter` appears to be working as expected!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## At which positions are amino acids different?\n",
    "The more interesting question is how these structures differ. We can use a similar for loop to address this question:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat and human are not the same at position 1, which is amino acid M\n",
      "cat and human are not the same at position 2, which is amino acid P\n",
      "cat and human are not the same at position 3, which is amino acid P\n",
      "cat and human are not the same at position 4, which is amino acid A\n",
      "cat and human are not the same at position 6, which is amino acid P\n",
      "cat and human are not the same at position 7, which is amino acid G\n",
      "cat and human are not the same at position 8, which is amino acid G\n",
      "cat and human are not the same at position 9, which is amino acid E\n",
      "cat and human are not the same at position 10, which is amino acid S\n",
      "cat and human are not the same at position 11, which is amino acid G\n",
      "cat and human are not the same at position 12, which is amino acid C\n",
      "cat and human are not the same at position 13, which is amino acid E\n",
      "cat and human are not the same at position 14, which is amino acid E\n",
      "cat and human are not the same at position 15, which is amino acid R\n",
      "cat and human are not the same at position 17, which is amino acid A\n",
      "cat and human are not the same at position 18, which is amino acid A\n",
      "cat and human are not the same at position 19, which is amino acid G\n",
      "cat and human are not the same at position 20, which is amino acid H\n",
      "cat and human are not the same at position 21, which is amino acid I\n",
      "cat and human are not the same at position 22, which is amino acid E\n",
      "cat and human are not the same at position 23, which is amino acid H\n",
      "cat and human are not the same at position 25, which is amino acid R\n",
      "cat and human are not the same at position 26, which is amino acid Y\n",
      "cat and human are not the same at position 27, which is amino acid L\n",
      "cat and human are not the same at position 28, which is amino acid S\n",
      "cat and human are not the same at position 29, which is amino acid L\n",
      "cat and human are not the same at position 30, which is amino acid L\n",
      "cat and human are not the same at position 31, which is amino acid D\n",
      "cat and human are not the same at position 32, which is amino acid A\n",
      "cat and human are not the same at position 33, which is amino acid V\n",
      "cat and human are not the same at position 34, which is amino acid D\n",
      "cat and human are not the same at position 35, which is amino acid N\n",
      "cat and human are not the same at position 36, which is amino acid S\n",
      "cat and human are not the same at position 37, which is amino acid K\n",
      "cat and human are not the same at position 38, which is amino acid M\n",
      "cat and human are not the same at position 39, which is amino acid A\n",
      "cat and human are not the same at position 40, which is amino acid L\n",
      "cat and human are not the same at position 41, which is amino acid N\n",
      "cat and human are not the same at position 42, which is amino acid S\n",
      "cat and human are not the same at position 43, which is amino acid G\n",
      "cat and human are not the same at position 44, which is amino acid S\n",
      "cat and human are not the same at position 45, which is amino acid P\n",
      "cat and human are not the same at position 46, which is amino acid P\n",
      "cat and human are not the same at position 48, which is amino acid I\n",
      "cat and human are not the same at position 49, which is amino acid G\n",
      "cat and human are not the same at position 50, which is amino acid P\n",
      "cat and human are not the same at position 51, which is amino acid Y\n",
      "cat and human are not the same at position 52, which is amino acid Y\n",
      "cat and human are not the same at position 53, which is amino acid E\n",
      "cat and human are not the same at position 54, which is amino acid N\n",
      "cat and human are not the same at position 55, which is amino acid H\n",
      "cat and human are not the same at position 56, which is amino acid G\n",
      "cat and human are not the same at position 57, which is amino acid Y\n",
      "cat and human are not the same at position 58, which is amino acid Q\n",
      "cat and human are not the same at position 59, which is amino acid P\n",
      "cat and human are not the same at position 60, which is amino acid E\n",
      "cat and human are not the same at position 61, which is amino acid N\n",
      "cat and human are not the same at position 62, which is amino acid P\n",
      "cat and human are not the same at position 63, which is amino acid Y\n",
      "cat and human are not the same at position 64, which is amino acid P\n",
      "cat and human are not the same at position 65, which is amino acid A\n",
      "cat and human are not the same at position 66, which is amino acid Q\n",
      "cat and human are not the same at position 67, which is amino acid P\n",
      "cat and human are not the same at position 68, which is amino acid T\n",
      "cat and human are not the same at position 69, which is amino acid V\n",
      "cat and human are not the same at position 70, which is amino acid V\n",
      "cat and human are not the same at position 71, which is amino acid P\n",
      "cat and human are not the same at position 72, which is amino acid T\n",
      "cat and human are not the same at position 73, which is amino acid V\n",
      "cat and human are not the same at position 74, which is amino acid Y\n",
      "cat and human are not the same at position 75, which is amino acid E\n",
      "cat and human are not the same at position 76, which is amino acid V\n",
      "cat and human are not the same at position 77, which is amino acid H\n",
      "cat and human are not the same at position 78, which is amino acid P\n",
      "cat and human are not the same at position 79, which is amino acid A\n",
      "cat and human are not the same at position 80, which is amino acid Q\n",
      "cat and human are not the same at position 81, which is amino acid Y\n",
      "cat and human are not the same at position 82, which is amino acid Y\n",
      "cat and human are not the same at position 83, which is amino acid P\n",
      "cat and human are not the same at position 84, which is amino acid S\n",
      "cat and human are not the same at position 85, which is amino acid P\n",
      "cat and human are not the same at position 86, which is amino acid V\n",
      "cat and human are not the same at position 87, which is amino acid P\n",
      "cat and human are not the same at position 88, which is amino acid Q\n",
      "cat and human are not the same at position 89, which is amino acid Y\n",
      "cat and human are not the same at position 90, which is amino acid A\n",
      "cat and human are not the same at position 91, which is amino acid P\n",
      "cat and human are not the same at position 92, which is amino acid R\n",
      "cat and human are not the same at position 93, which is amino acid V\n",
      "cat and human are not the same at position 94, which is amino acid L\n",
      "cat and human are not the same at position 95, which is amino acid T\n",
      "cat and human are not the same at position 96, which is amino acid Q\n",
      "cat and human are not the same at position 97, which is amino acid A\n",
      "cat and human are not the same at position 98, which is amino acid S\n",
      "cat and human are not the same at position 99, which is amino acid N\n",
      "cat and human are not the same at position 100, which is amino acid P\n",
      "cat and human are not the same at position 101, which is amino acid V\n",
      "cat and human are not the same at position 102, which is amino acid V\n",
      "cat and human are not the same at position 103, which is amino acid C\n",
      "cat and human are not the same at position 104, which is amino acid T\n",
      "cat and human are not the same at position 105, which is amino acid Q\n",
      "cat and human are not the same at position 107, which is amino acid K\n",
      "cat and human are not the same at position 108, which is amino acid S\n",
      "cat and human are not the same at position 109, which is amino acid P\n",
      "cat and human are not the same at position 110, which is amino acid S\n",
      "cat and human are not the same at position 111, which is amino acid G\n",
      "cat and human are not the same at position 112, which is amino acid T\n",
      "cat and human are not the same at position 113, which is amino acid V\n",
      "cat and human are not the same at position 114, which is amino acid C\n",
      "cat and human are not the same at position 115, which is amino acid T\n",
      "cat and human are not the same at position 116, which is amino acid S\n",
      "cat and human are not the same at position 117, which is amino acid K\n",
      "cat and human are not the same at position 118, which is amino acid T\n",
      "cat and human are not the same at position 119, which is amino acid K\n",
      "cat and human are not the same at position 120, which is amino acid K\n",
      "cat and human are not the same at position 121, which is amino acid A\n",
      "cat and human are not the same at position 122, which is amino acid L\n",
      "cat and human are not the same at position 123, which is amino acid C\n",
      "cat and human are not the same at position 124, which is amino acid I\n",
      "cat and human are not the same at position 125, which is amino acid T\n",
      "cat and human are not the same at position 126, which is amino acid L\n",
      "cat and human are not the same at position 127, which is amino acid T\n",
      "cat and human are not the same at position 128, which is amino acid L\n",
      "cat and human are not the same at position 129, which is amino acid G\n",
      "cat and human are not the same at position 130, which is amino acid T\n",
      "cat and human are not the same at position 131, which is amino acid F\n",
      "cat and human are not the same at position 132, which is amino acid L\n",
      "cat and human are not the same at position 133, which is amino acid V\n",
      "cat and human are not the same at position 134, which is amino acid G\n",
      "cat and human are not the same at position 135, which is amino acid A\n",
      "cat and human are not the same at position 136, which is amino acid A\n",
      "cat and human are not the same at position 137, which is amino acid L\n",
      "cat and human are not the same at position 138, which is amino acid A\n",
      "cat and human are not the same at position 139, which is amino acid A\n",
      "cat and human are not the same at position 140, which is amino acid G\n",
      "cat and human are not the same at position 141, which is amino acid L\n",
      "cat and human are not the same at position 142, which is amino acid L\n",
      "cat and human are not the same at position 143, which is amino acid W\n",
      "cat and human are not the same at position 144, which is amino acid K\n",
      "cat and human are not the same at position 145, which is amino acid F\n",
      "cat and human are not the same at position 146, which is amino acid M\n",
      "cat and human are not the same at position 147, which is amino acid G\n",
      "cat and human are not the same at position 148, which is amino acid S\n",
      "cat and human are not the same at position 149, which is amino acid K\n",
      "cat and human are not the same at position 150, which is amino acid C\n",
      "cat and human are not the same at position 151, which is amino acid S\n",
      "cat and human are not the same at position 152, which is amino acid N\n",
      "cat and human are not the same at position 153, which is amino acid S\n",
      "cat and human are not the same at position 154, which is amino acid G\n",
      "cat and human are not the same at position 155, which is amino acid I\n",
      "cat and human are not the same at position 156, which is amino acid E\n",
      "cat and human are not the same at position 158, which is amino acid D\n",
      "cat and human are not the same at position 159, which is amino acid S\n",
      "cat and human are not the same at position 160, which is amino acid S\n",
      "cat and human are not the same at position 161, which is amino acid G\n",
      "cat and human are not the same at position 162, which is amino acid T\n",
      "cat and human are not the same at position 163, which is amino acid C\n",
      "cat and human are not the same at position 164, which is amino acid I\n",
      "cat and human are not the same at position 165, which is amino acid N\n",
      "cat and human are not the same at position 166, which is amino acid P\n",
      "cat and human are not the same at position 167, which is amino acid S\n",
      "cat and human are not the same at position 168, which is amino acid N\n",
      "cat and human are not the same at position 169, which is amino acid W\n",
      "cat and human are not the same at position 171, which is amino acid D\n",
      "cat and human are not the same at position 173, which is amino acid V\n",
      "cat and human are not the same at position 174, which is amino acid S\n",
      "cat and human are not the same at position 175, which is amino acid H\n",
      "cat and human are not the same at position 177, which is amino acid P\n",
      "cat and human are not the same at position 178, which is amino acid G\n",
      "cat and human are not the same at position 180, which is amino acid E\n",
      "cat and human are not the same at position 181, which is amino acid D\n",
      "cat and human are not the same at position 182, which is amino acid E\n",
      "cat and human are not the same at position 183, which is amino acid N\n",
      "cat and human are not the same at position 184, which is amino acid R\n",
      "cat and human are not the same at position 185, which is amino acid C\n",
      "cat and human are not the same at position 186, which is amino acid V\n",
      "cat and human are not the same at position 187, which is amino acid R\n",
      "cat and human are not the same at position 188, which is amino acid L\n",
      "cat and human are not the same at position 189, which is amino acid Y\n",
      "cat and human are not the same at position 190, which is amino acid G\n",
      "cat and human are not the same at position 191, which is amino acid P\n",
      "cat and human are not the same at position 192, which is amino acid N\n",
      "cat and human are not the same at position 193, which is amino acid F\n",
      "cat and human are not the same at position 194, which is amino acid I\n",
      "cat and human are not the same at position 195, which is amino acid L\n",
      "cat and human are not the same at position 196, which is amino acid Q\n",
      "cat and human are not the same at position 197, which is amino acid V\n",
      "cat and human are not the same at position 198, which is amino acid Y\n",
      "cat and human are not the same at position 199, which is amino acid S\n",
      "cat and human are not the same at position 200, which is amino acid S\n",
      "cat and human are not the same at position 201, which is amino acid Q\n",
      "cat and human are not the same at position 202, which is amino acid R\n",
      "cat and human are not the same at position 203, which is amino acid K\n",
      "cat and human are not the same at position 204, which is amino acid S\n",
      "cat and human are not the same at position 205, which is amino acid W\n",
      "cat and human are not the same at position 206, which is amino acid H\n",
      "cat and human are not the same at position 207, which is amino acid P\n",
      "cat and human are not the same at position 208, which is amino acid V\n",
      "cat and human are not the same at position 209, which is amino acid C\n",
      "cat and human are not the same at position 210, which is amino acid Q\n",
      "cat and human are not the same at position 211, which is amino acid D\n",
      "cat and human are not the same at position 212, which is amino acid D\n",
      "cat and human are not the same at position 213, which is amino acid W\n",
      "cat and human are not the same at position 214, which is amino acid N\n",
      "cat and human are not the same at position 215, which is amino acid E\n",
      "cat and human are not the same at position 216, which is amino acid N\n",
      "cat and human are not the same at position 217, which is amino acid Y\n",
      "cat and human are not the same at position 218, which is amino acid G\n",
      "cat and human are not the same at position 219, which is amino acid R\n",
      "cat and human are not the same at position 220, which is amino acid A\n",
      "cat and human are not the same at position 221, which is amino acid A\n",
      "cat and human are not the same at position 223, which is amino acid R\n",
      "cat and human are not the same at position 224, which is amino acid D\n",
      "cat and human are not the same at position 225, which is amino acid M\n",
      "cat and human are not the same at position 226, which is amino acid G\n",
      "cat and human are not the same at position 227, which is amino acid Y\n",
      "cat and human are not the same at position 228, which is amino acid K\n",
      "cat and human are not the same at position 229, which is amino acid N\n",
      "cat and human are not the same at position 230, which is amino acid N\n",
      "cat and human are not the same at position 231, which is amino acid F\n",
      "cat and human are not the same at position 232, which is amino acid Y\n",
      "cat and human are not the same at position 233, which is amino acid S\n",
      "cat and human are not the same at position 234, which is amino acid S\n",
      "cat and human are not the same at position 235, which is amino acid Q\n",
      "cat and human are not the same at position 236, which is amino acid G\n",
      "cat and human are not the same at position 237, which is amino acid I\n",
      "cat and human are not the same at position 238, which is amino acid V\n",
      "cat and human are not the same at position 239, which is amino acid D\n",
      "cat and human are not the same at position 240, which is amino acid D\n",
      "cat and human are not the same at position 241, which is amino acid S\n",
      "cat and human are not the same at position 242, which is amino acid G\n",
      "cat and human are not the same at position 243, which is amino acid S\n",
      "cat and human are not the same at position 244, which is amino acid T\n",
      "cat and human are not the same at position 245, which is amino acid S\n",
      "cat and human are not the same at position 246, which is amino acid F\n",
      "cat and human are not the same at position 247, which is amino acid M\n",
      "cat and human are not the same at position 248, which is amino acid K\n",
      "cat and human are not the same at position 249, which is amino acid L\n",
      "cat and human are not the same at position 250, which is amino acid N\n",
      "cat and human are not the same at position 251, which is amino acid T\n",
      "cat and human are not the same at position 252, which is amino acid S\n",
      "cat and human are not the same at position 253, which is amino acid A\n",
      "cat and human are not the same at position 254, which is amino acid G\n",
      "cat and human are not the same at position 255, which is amino acid N\n",
      "cat and human are not the same at position 256, which is amino acid V\n",
      "cat and human are not the same at position 257, which is amino acid D\n",
      "cat and human are not the same at position 258, which is amino acid I\n",
      "cat and human are not the same at position 259, which is amino acid Y\n",
      "cat and human are not the same at position 260, which is amino acid K\n",
      "cat and human are not the same at position 261, which is amino acid K\n",
      "cat and human are not the same at position 262, which is amino acid L\n",
      "cat and human are not the same at position 263, which is amino acid Y\n",
      "cat and human are not the same at position 264, which is amino acid H\n",
      "cat and human are not the same at position 265, which is amino acid S\n",
      "cat and human are not the same at position 266, which is amino acid D\n",
      "cat and human are not the same at position 267, which is amino acid A\n",
      "cat and human are not the same at position 268, which is amino acid C\n",
      "cat and human are not the same at position 269, which is amino acid S\n",
      "cat and human are not the same at position 270, which is amino acid S\n",
      "cat and human are not the same at position 271, which is amino acid K\n",
      "cat and human are not the same at position 272, which is amino acid A\n",
      "cat and human are not the same at position 273, which is amino acid V\n",
      "cat and human are not the same at position 274, which is amino acid V\n",
      "cat and human are not the same at position 275, which is amino acid S\n",
      "cat and human are not the same at position 276, which is amino acid L\n",
      "cat and human are not the same at position 277, which is amino acid R\n",
      "cat and human are not the same at position 278, which is amino acid C\n",
      "cat and human are not the same at position 279, which is amino acid I\n",
      "cat and human are not the same at position 280, which is amino acid A\n",
      "cat and human are not the same at position 283, which is amino acid V\n",
      "cat and human are not the same at position 284, which is amino acid N\n",
      "cat and human are not the same at position 285, which is amino acid L\n",
      "cat and human are not the same at position 286, which is amino acid N\n",
      "cat and human are not the same at position 287, which is amino acid S\n",
      "cat and human are not the same at position 288, which is amino acid S\n",
      "cat and human are not the same at position 289, which is amino acid R\n",
      "cat and human are not the same at position 290, which is amino acid Q\n",
      "cat and human are not the same at position 291, which is amino acid S\n",
      "cat and human are not the same at position 294, which is amino acid V\n",
      "cat and human are not the same at position 297, which is amino acid E\n",
      "cat and human are not the same at position 298, which is amino acid S\n",
      "cat and human are not the same at position 299, which is amino acid A\n",
      "cat and human are not the same at position 300, which is amino acid L\n",
      "cat and human are not the same at position 301, which is amino acid P\n",
      "cat and human are not the same at position 303, which is amino acid A\n",
      "cat and human are not the same at position 304, which is amino acid W\n",
      "cat and human are not the same at position 308, which is amino acid V\n",
      "cat and human are not the same at position 309, which is amino acid S\n",
      "cat and human are not the same at position 311, which is amino acid H\n",
      "cat and human are not the same at position 312, which is amino acid V\n",
      "cat and human are not the same at position 313, which is amino acid Q\n",
      "cat and human are not the same at position 314, which is amino acid N\n",
      "cat and human are not the same at position 315, which is amino acid V\n",
      "cat and human are not the same at position 317, which is amino acid V\n",
      "cat and human are not the same at position 320, which is amino acid G\n",
      "cat and human are not the same at position 321, which is amino acid S\n",
      "cat and human are not the same at position 322, which is amino acid I\n",
      "cat and human are not the same at position 323, which is amino acid I\n",
      "cat and human are not the same at position 324, which is amino acid T\n",
      "cat and human are not the same at position 325, which is amino acid P\n",
      "cat and human are not the same at position 326, which is amino acid E\n",
      "cat and human are not the same at position 328, which is amino acid I\n",
      "cat and human are not the same at position 329, which is amino acid V\n",
      "cat and human are not the same at position 330, which is amino acid T\n",
      "cat and human are not the same at position 335, which is amino acid V\n",
      "cat and human are not the same at position 336, which is amino acid E\n",
      "cat and human are not the same at position 337, which is amino acid K\n",
      "cat and human are not the same at position 338, which is amino acid P\n",
      "cat and human are not the same at position 339, which is amino acid L\n",
      "cat and human are not the same at position 340, which is amino acid N\n",
      "cat and human are not the same at position 341, which is amino acid N\n",
      "cat and human are not the same at position 342, which is amino acid P\n",
      "cat and human are not the same at position 343, which is amino acid W\n",
      "cat and human are not the same at position 344, which is amino acid H\n",
      "cat and human are not the same at position 345, which is amino acid W\n",
      "cat and human are not the same at position 346, which is amino acid T\n",
      "cat and human are not the same at position 347, which is amino acid A\n",
      "cat and human are not the same at position 348, which is amino acid F\n",
      "cat and human are not the same at position 349, which is amino acid A\n",
      "cat and human are not the same at position 351, which is amino acid I\n",
      "cat and human are not the same at position 352, which is amino acid L\n",
      "cat and human are not the same at position 353, which is amino acid R\n",
      "cat and human are not the same at position 354, which is amino acid Q\n",
      "cat and human are not the same at position 355, which is amino acid S\n",
      "cat and human are not the same at position 356, which is amino acid F\n",
      "cat and human are not the same at position 357, which is amino acid M\n",
      "cat and human are not the same at position 358, which is amino acid F\n",
      "cat and human are not the same at position 359, which is amino acid Y\n",
      "cat and human are not the same at position 360, which is amino acid G\n",
      "cat and human are not the same at position 361, which is amino acid A\n",
      "cat and human are not the same at position 362, which is amino acid G\n",
      "cat and human are not the same at position 363, which is amino acid Y\n",
      "cat and human are not the same at position 364, which is amino acid Q\n",
      "cat and human are not the same at position 365, which is amino acid V\n",
      "cat and human are not the same at position 366, which is amino acid E\n",
      "cat and human are not the same at position 367, which is amino acid K\n",
      "cat and human are not the same at position 368, which is amino acid V\n",
      "cat and human are not the same at position 369, which is amino acid I\n",
      "cat and human are not the same at position 370, which is amino acid S\n",
      "cat and human are not the same at position 372, which is amino acid P\n",
      "cat and human are not the same at position 373, which is amino acid N\n",
      "cat and human are not the same at position 374, which is amino acid Y\n",
      "cat and human are not the same at position 375, which is amino acid D\n",
      "cat and human are not the same at position 376, which is amino acid S\n",
      "cat and human are not the same at position 377, which is amino acid K\n",
      "cat and human are not the same at position 378, which is amino acid T\n",
      "cat and human are not the same at position 379, which is amino acid K\n",
      "cat and human are not the same at position 380, which is amino acid N\n",
      "cat and human are not the same at position 381, which is amino acid N\n",
      "cat and human are not the same at position 386, which is amino acid M\n",
      "cat and human are not the same at position 387, which is amino acid K\n",
      "cat and human are not the same at position 389, which is amino acid Q\n",
      "cat and human are not the same at position 390, which is amino acid K\n",
      "cat and human are not the same at position 391, which is amino acid P\n",
      "cat and human are not the same at position 392, which is amino acid L\n",
      "cat and human are not the same at position 393, which is amino acid T\n",
      "cat and human are not the same at position 394, which is amino acid F\n",
      "cat and human are not the same at position 395, which is amino acid N\n",
      "cat and human are not the same at position 396, which is amino acid D\n",
      "cat and human are not the same at position 397, which is amino acid L\n",
      "cat and human are not the same at position 398, which is amino acid V\n",
      "cat and human are not the same at position 399, which is amino acid K\n",
      "cat and human are not the same at position 401, which is amino acid V\n",
      "cat and human are not the same at position 405, which is amino acid N\n",
      "cat and human are not the same at position 406, which is amino acid P\n",
      "cat and human are not the same at position 407, which is amino acid G\n",
      "cat and human are not the same at position 408, which is amino acid M\n",
      "cat and human are not the same at position 409, which is amino acid M\n",
      "cat and human are not the same at position 410, which is amino acid L\n",
      "cat and human are not the same at position 411, which is amino acid Q\n",
      "cat and human are not the same at position 412, which is amino acid P\n",
      "cat and human are not the same at position 413, which is amino acid E\n",
      "cat and human are not the same at position 414, which is amino acid Q\n",
      "cat and human are not the same at position 415, which is amino acid L\n",
      "cat and human are not the same at position 417, which is amino acid W\n",
      "cat and human are not the same at position 418, which is amino acid I\n",
      "cat and human are not the same at position 419, which is amino acid S\n",
      "cat and human are not the same at position 423, which is amino acid A\n",
      "cat and human are not the same at position 424, which is amino acid T\n",
      "cat and human are not the same at position 425, which is amino acid E\n",
      "cat and human are not the same at position 426, which is amino acid E\n",
      "cat and human are not the same at position 427, which is amino acid K\n",
      "cat and human are not the same at position 428, which is amino acid G\n",
      "cat and human are not the same at position 429, which is amino acid K\n",
      "cat and human are not the same at position 430, which is amino acid T\n",
      "cat and human are not the same at position 431, which is amino acid S\n",
      "cat and human are not the same at position 432, which is amino acid E\n",
      "cat and human are not the same at position 433, which is amino acid V\n",
      "cat and human are not the same at position 435, which is amino acid N\n",
      "cat and human are not the same at position 436, which is amino acid A\n",
      "cat and human are not the same at position 437, which is amino acid A\n",
      "cat and human are not the same at position 438, which is amino acid K\n",
      "cat and human are not the same at position 439, which is amino acid V\n",
      "cat and human are not the same at position 440, which is amino acid L\n",
      "cat and human are not the same at position 441, which is amino acid L\n",
      "cat and human are not the same at position 442, which is amino acid I\n",
      "cat and human are not the same at position 443, which is amino acid E\n",
      "cat and human are not the same at position 444, which is amino acid T\n",
      "cat and human are not the same at position 445, which is amino acid Q\n",
      "cat and human are not the same at position 446, which is amino acid R\n",
      "cat and human are not the same at position 448, which is amino acid N\n",
      "cat and human are not the same at position 449, which is amino acid S\n",
      "cat and human are not the same at position 450, which is amino acid R\n",
      "cat and human are not the same at position 451, which is amino acid Y\n",
      "cat and human are not the same at position 452, which is amino acid V\n",
      "cat and human are not the same at position 454, which is amino acid D\n",
      "cat and human are not the same at position 455, which is amino acid N\n",
      "cat and human are not the same at position 456, which is amino acid L\n",
      "cat and human are not the same at position 457, which is amino acid I\n",
      "cat and human are not the same at position 458, which is amino acid T\n",
      "cat and human are not the same at position 459, which is amino acid P\n",
      "cat and human are not the same at position 460, which is amino acid A\n",
      "cat and human are not the same at position 462, which is amino acid I\n",
      "cat and human are not the same at position 466, which is amino acid F\n",
      "cat and human are not the same at position 467, which is amino acid L\n",
      "cat and human are not the same at position 468, which is amino acid Q\n",
      "cat and human are not the same at position 470, which is amino acid N\n",
      "cat and human are not the same at position 471, which is amino acid V\n",
      "cat and human are not the same at position 473, which is amino acid S\n",
      "cat and human are not the same at position 475, which is amino acid Q\n",
      "cat and human are not the same at position 483, which is amino acid V\n",
      "cat and human are not the same at position 484, which is amino acid T\n",
      "cat and human are not the same at position 485, which is amino acid S\n",
      "cat and human are not the same at position 486, which is amino acid K\n",
      "cat and human are not the same at position 487, which is amino acid N\n",
      "cat and human are not the same at position 488, which is amino acid N\n",
      "cat and human are not the same at position 489, which is amino acid I\n",
      "cat and human are not the same at position 491, which is amino acid W\n",
      "cat and human are not the same at position 493, which is amino acid I\n",
      "cat and human are not the same at position 495, which is amino acid D\n",
      "cat and human are not the same at position 496, which is amino acid T\n",
      "cat and human are not the same at position 500, which is amino acid S\n",
      "cat and human are not the same at position 501, which is amino acid G\n",
      "cat and human are not the same at position 503, which is amino acid A\n",
      "cat and human are not the same at position 504, which is amino acid K\n",
      "cat and human are not the same at position 505, which is amino acid A\n",
      "cat and human are not the same at position 506, which is amino acid Y\n",
      "cat and human are not the same at position 507, which is amino acid R\n",
      "cat and human are not the same at position 512, which is amino acid G\n",
      "cat and human are not the same at position 513, which is amino acid N\n",
      "cat and human are not the same at position 515, which is amino acid M\n",
      "cat and human are not the same at position 516, which is amino acid V\n",
      "cat and human are not the same at position 517, which is amino acid F\n",
      "cat and human are not the same at position 518, which is amino acid T\n",
      "cat and human are not the same at position 519, which is amino acid D\n",
      "cat and human are not the same at position 522, which is amino acid Y\n",
      "cat and human are not the same at position 523, which is amino acid R\n",
      "cat and human are not the same at position 524, which is amino acid Q\n",
      "cat and human are not the same at position 525, which is amino acid M\n",
      "cat and human are not the same at position 526, which is amino acid R\n",
      "cat and human are not the same at position 527, which is amino acid A\n",
      "cat and human are not the same at position 528, which is amino acid D\n",
      "cat and human are not the same at position 529, which is amino acid G\n"
     ]
    }
   ],
   "source": [
    "# we want to keep track of which amino acid our\n",
    "# \"cursor\" is on in the for loop\n",
    "position_counter = 0\n",
    "\n",
    "list_to_store_same = list()\n",
    "\n",
    "for position in hum_cat_pssm.pssm:\n",
    "    \n",
    "    # `position` is the 2-element tuple\n",
    "    # let's give each element a useful name\n",
    "    resi_in_human = position[0]\n",
    "    resi_dict = position[1]\n",
    "    \n",
    "    # skip this position if it is a '-'\n",
    "    # in the human sequence record\n",
    "    if resi_in_human == '-':\n",
    "        continue\n",
    "    else:\n",
    "        # increment the counter by 1\n",
    "        position_counter += 1\n",
    "    \n",
    "    # if more than one instance of amino acid\n",
    "    # `resi_in_human` was found at this position,\n",
    "    # meaning that the cat homolog is the same amino acid\n",
    "    if position[1][resi_in_human] != 2:\n",
    "        print(f\"cat and human are the same at position \" +\n",
    "            f\"{position_counter}, which is amino acid {resi_in_human}\")\n",
    "        # list_to_store_same.append(position_counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## At which positions do we encounter a hydrophobic -> hydrophilic (or vice versa)?\n",
    "For this question, we will need to make our algorithm a little more complex. We are going to start by making a dataframe that stores amino acid properties, such as volume, hydrophobicity, charge, and so forth. We will use the CSV format of this [table of amino acid properties](https://web.nmsu.edu/~talipovm/lib/exe/fetch.php?media=world:pasted:table08.pdf) and load it into a [Pandas DataFrame](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html#object-creation).\n",
    "\n",
    "Let's also narrow our analysis to sites that are implicated as important in cleaving the SARS-CoV-2 S protein. H296, D345 and S441 are the catalytic triad, and D435 is a binding residue ([Meng et al 2020](https://www.biorxiv.org/content/10.1101/2020.02.08.926006v3.full)).\n",
    "\n",
    "K225 is implicated as important in binding monobasic targets such as S1/S2 domain of S protein ([Ohno et al 2020](https://www.biorxiv.org/content/10.1101/2020.06.12.149229v1.full)). Residue 225 in isoform 1 is actually a Leucine (L); this might have been a typo, since the two previous residues (223 and 224) are both lysines. We will consider both 223 and 224 as important, since they likely both contribute to the positive patch in the binding site, hypothesized to confer preference for monobasic substrates by Ohno et al."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "M\n"
     ]
    }
   ],
   "source": [
    "for position in human_compact:\n",
    "    print(type(position))\n",
    "    print(position[0])\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pseudocode outline\n",
    "\n",
    "```\n",
    "human MP   P   APP\n",
    "cat   LA   P   ---\n",
    "```\n",
    "\n",
    "1. Iterate over each amino acid. For loop. We don't necessarily need the PSSM here.\n",
    "\n",
    "```python\n",
    "for position in human_compact:\n",
    "```\n",
    "2. Get the amino acid at this position, for both human and cat.\n",
    "\n",
    "```python\n",
    "resi_in_human = human_compact[position]\n",
    "resi_in_cat = cat_compact[position]\n",
    "```\n",
    "3. Get the hydrophobicity of each amino acid (5.0 and 3.0)\n",
    "\n",
    "```python\n",
    "h_hum = get_hydrophobicity(resi_in_human)\n",
    "h_cat = get_hydrophobicity(resi_in_cat)\n",
    "```\n",
    "4. Get the (absolute value of) **difference** in hydrophobicity (2.0)\n",
    "\n",
    "```python\n",
    "h_hum = 9.00\n",
    "h_cat = 20.0\n",
    "diff = abs(h_hum - h_cat)\n",
    "diff = 11.0\n",
    "```\n",
    "5. Is this difference \"large\" == 5.0 -> yes or no (boolean, True or False)\n",
    "\n",
    "```python\n",
    "if diff < 5.0:\n",
    "    # not a change in hydrophobicity\n",
    "    no_change_in_h.append(position_counter)\n",
    "else:\n",
    "    # is a change in hydrophobicity\n",
    "    change_in_h.append(position_counter)\n",
    "\n",
    "position_counter += 1\n",
    "```\n",
    "6. Variable that stores this boolean\n",
    "    1. List that has length of human sequence\n",
    "    2. Value is this boolean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "change_in_h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "single_letter\n",
       "A     9.0\n",
       "R    15.0\n",
       "N    16.0\n",
       "D    19.0\n",
       "C     7.0\n",
       "Q    17.0\n",
       "E    18.0\n",
       "G    11.0\n",
       "H    10.0\n",
       "I     1.0\n",
       "L     3.0\n",
       "K    20.0\n",
       "M     5.0\n",
       "F     2.0\n",
       "P    13.0\n",
       "S    14.0\n",
       "T    12.0\n",
       "W     6.0\n",
       "Y     8.0\n",
       "V     4.0\n",
       "Name: hydrophobicity, dtype: float64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_props['hydrophobicity']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\".replace(\"-\", '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hydrophobicity(aa):\n",
    "    hydrophobicity = aa_props.loc[[aa]]['hydrophobicity'].item()\n",
    "    return hydrophobicity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14.0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa = 'S'\n",
    "get_hydrophobicity(aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_name</th>\n",
       "      <th>pka</th>\n",
       "      <th>mass</th>\n",
       "      <th>occurrence</th>\n",
       "      <th>percent_buried</th>\n",
       "      <th>buried_vol</th>\n",
       "      <th>vdw</th>\n",
       "      <th>accessible</th>\n",
       "      <th>hydrophobicity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>single_letter</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>alanine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.08</td>\n",
       "      <td>7.5</td>\n",
       "      <td>38.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>argenine</td>\n",
       "      <td>12.5</td>\n",
       "      <td>156.10</td>\n",
       "      <td>5.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>225.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>asparagine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>114.04</td>\n",
       "      <td>4.6</td>\n",
       "      <td>10.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>aspartic acid</td>\n",
       "      <td>3.9</td>\n",
       "      <td>115.08</td>\n",
       "      <td>5.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>125.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>cysteine</td>\n",
       "      <td>8.3</td>\n",
       "      <td>103.01</td>\n",
       "      <td>1.8</td>\n",
       "      <td>47.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>glutamine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128.06</td>\n",
       "      <td>4.1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>161.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>glutamic acid</td>\n",
       "      <td>4.3</td>\n",
       "      <td>129.12</td>\n",
       "      <td>6.3</td>\n",
       "      <td>20.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>glycine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.05</td>\n",
       "      <td>7.1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>histidine</td>\n",
       "      <td>6.0</td>\n",
       "      <td>137.16</td>\n",
       "      <td>2.2</td>\n",
       "      <td>19.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>isoleucine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.16</td>\n",
       "      <td>5.5</td>\n",
       "      <td>65.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>leucine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.16</td>\n",
       "      <td>9.1</td>\n",
       "      <td>41.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>lysine</td>\n",
       "      <td>10.8</td>\n",
       "      <td>128.17</td>\n",
       "      <td>5.8</td>\n",
       "      <td>4.2</td>\n",
       "      <td>171.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>methionine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>131.20</td>\n",
       "      <td>2.8</td>\n",
       "      <td>50.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>phenylalanine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147.18</td>\n",
       "      <td>3.9</td>\n",
       "      <td>48.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>proline</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.12</td>\n",
       "      <td>5.1</td>\n",
       "      <td>24.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>serine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87.08</td>\n",
       "      <td>7.4</td>\n",
       "      <td>24.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>threonine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>101.11</td>\n",
       "      <td>6.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>tryptophan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>186.21</td>\n",
       "      <td>1.3</td>\n",
       "      <td>23.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>tyrosine</td>\n",
       "      <td>10.9</td>\n",
       "      <td>163.06</td>\n",
       "      <td>3.3</td>\n",
       "      <td>13.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>valine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.13</td>\n",
       "      <td>6.5</td>\n",
       "      <td>56.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   full_name   pka    mass  occurrence  percent_buried  \\\n",
       "single_letter                                                            \n",
       "A                    alanine   NaN   71.08         7.5            38.0   \n",
       "R                   argenine  12.5  156.10         5.2             0.0   \n",
       "N                 asparagine   NaN  114.04         4.6            10.0   \n",
       "D              aspartic acid   3.9  115.08         5.2            14.5   \n",
       "C                   cysteine   8.3  103.01         1.8            47.0   \n",
       "Q                  glutamine   NaN  128.06         4.1             6.3   \n",
       "E              glutamic acid   4.3  129.12         6.3            20.0   \n",
       "G                    glycine   NaN   57.05         7.1            37.0   \n",
       "H                  histidine   6.0  137.16         2.2            19.0   \n",
       "I                 isoleucine   NaN  113.16         5.5            65.0   \n",
       "L                    leucine   NaN  113.16         9.1            41.0   \n",
       "K                     lysine  10.8  128.17         5.8             4.2   \n",
       "M                 methionine   NaN  131.20         2.8            50.0   \n",
       "F              phenylalanine   NaN  147.18         3.9            48.0   \n",
       "P                    proline   NaN   97.12         5.1            24.0   \n",
       "S                     serine   NaN   87.08         7.4            24.0   \n",
       "T                  threonine   NaN  101.11         6.0            25.0   \n",
       "W                 tryptophan   NaN  186.21         1.3            23.0   \n",
       "Y                   tyrosine  10.9  163.06         3.3            13.0   \n",
       "V                     valine   NaN   99.13         6.5            56.0   \n",
       "\n",
       "               buried_vol    vdw  accessible  hydrophobicity  \n",
       "single_letter                                                 \n",
       "A                    92.0   67.0        67.0             9.0  \n",
       "R                   225.0  148.0       196.0            15.0  \n",
       "N                   135.0   96.0       113.0            16.0  \n",
       "D                   125.0   91.0       106.0            19.0  \n",
       "C                   106.0   86.0       104.0             7.0  \n",
       "Q                   161.0  114.0       144.0            17.0  \n",
       "E                   155.0  109.0       138.0            18.0  \n",
       "G                    66.0   48.0         NaN            11.0  \n",
       "H                   167.0  118.0       151.0            10.0  \n",
       "I                   169.0  124.0       140.0             1.0  \n",
       "L                   168.0  124.0       137.0             3.0  \n",
       "K                   171.0  135.0       167.0            20.0  \n",
       "M                   171.0  124.0       160.0             5.0  \n",
       "F                   203.0  135.0       175.0             2.0  \n",
       "P                   129.0   90.0       105.0            13.0  \n",
       "S                    99.0   73.0        80.0            14.0  \n",
       "T                   122.0   93.0       102.0            12.0  \n",
       "W                   240.0  163.0       217.0             6.0  \n",
       "Y                   203.0  141.0       187.0             8.0  \n",
       "V                   142.0  105.0       117.0             4.0  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_props = pd.read_csv(\"../../data/amino_acid_properties.csv\")\n",
    "aa_props.set_index('single_letter', inplace=True)\n",
    "aa_props"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See the [PDF format](../../data/amino_acid_properties.pdf) for references and details on how these metrics are calculated.\n",
    "\n",
    "Next, we will write a Python [function](https://github.com/wilfredinni/python-cheatsheet#functions), in which we pass the single-letter IDs of two amino acids, and get a Python [boolean](https://github.com/wilfredinni/python-cheatsheet#boolean-operators) (variable that stores `True` or `False`) that says whether or not these two amino acids have different hydrophobicity. We arbitrarily define \"difference in hydrophobicity\" here as a difference of 5.0 units between the amino acids' `hydrophobicity` columns.\n",
    "\n",
    "The text at the beginning of the funciton that is wrapped in `\"\"\"` is a special type of [comment](https://github.com/wilfredinni/python-cheatsheet#comments) called a [function docstring](https://github.com/wilfredinni/python-cheatsheet#comments); it tells us what the function does and how to use it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_change_in_hydrophobicity(resi1, resi2):\n",
    "    \"\"\"This function takes string-type amino acid identifiers `resi1` and `resi2`\n",
    "    and compares their hydrophobicities. If the absolute value of the difference\n",
    "    between hydrophobicities is greater than `min_diff`, return boolean True.\n",
    "    Otherwise, return boolean False.\n",
    "    \"\"\"\n",
    "    min_diff = 5.0\n",
    "    print(f\"comparing hydrophobicity between {resi1} and {resi2}\")\n",
    "    h1 = aa_props.loc[[resi1]]['hydrophobicity'].item()\n",
    "    h2 = aa_props.loc[[resi2]]['hydrophobicity'].item()\n",
    "    \n",
    "    diff = abs(h1 - h2)\n",
    "    print(f\"the difference is hydrophobicity is {diff}\")\n",
    "    \n",
    "    if diff > min_diff:\n",
    "        return True\n",
    "    else:\n",
    "        return False    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can quickly test our function with some examples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "comparing hydrophobicity between M and S\n",
      "the difference is hydrophobicity is 9.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_change_in_hydrophobicity('M', 'S')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "comparing hydrophobicity between M and F\n",
      "the difference is hydrophobicity is 3.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_change_in_hydrophobicity('M', 'F')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "comparing hydrophobicity between M and M\n",
      "the difference is hydrophobicity is 0.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_change_in_hydrophobicity('M', 'M')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get list of interesting residues\n",
    "\n",
    "Next, let's generate a list of positions in the human sequence that are residues of interest, such as the catalytic triad (H296, D345 and S441) and important binding residues (D435, K223, and K224).\n",
    "\n",
    "It is important to remember that these positions reported in the literature are relative to the human **isoform 1** sequence, not the **isoform 2** sequence (which we have stored in the variable `human_compact`). Thankfully, the conversion it relatively simple: isoform 2 is simply a splice variant in which `M → MPPAPPGGESGCEERGAAGHIEHSRYLSLLDAVDNSKM` at the N-terminal methionine. This means that we simply add 37 to the isoform 1 index to get the isoform 2 index. For instance, the catalytic serine S441 in isoform 1 is at position 441 + 37 = 478 in isoform 2.\n",
    "\n",
    "Lastly, amino acid numbering in the literature uses 1 indexing (first amino acid is `M`), while our Python sequence uses 0 indexing. So position 478 with 1-indexing can be indexed using 477 with 0-indexing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "529"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(str(human_compact.seq).replace('-', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'S'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(human_compact.seq).replace('-', '')[477]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also check that the other residues of interest are the expected amino acids:\n",
    "* H296 in isoform 1 → 296 + 37 = H333 in isoform 2 → 333 - 1 = position 332 with 0-indexing\n",
    "* D345 → 381\n",
    "* D435 → 471\n",
    "* K223 → 259\n",
    "* K224 → 260\n",
    "\n",
    "Let's store these 0-indexed positions in a list so we can use it later:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "resi_interest = [332, 381, 471, 259, 260]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check that these positions are the amino acids we expect, this time using a for loop:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "amino acid at 0-indexed position 332 is H\n",
      "amino acid at 0-indexed position 381 is D\n",
      "amino acid at 0-indexed position 471 is D\n",
      "amino acid at 0-indexed position 259 is K\n",
      "amino acid at 0-indexed position 260 is K\n"
     ]
    }
   ],
   "source": [
    "for position in resi_interest:\n",
    "    resi = str(human_compact.seq).replace('-', '')[position]\n",
    "    print(f\"amino acid at 0-indexed position {position} is {resi}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Putting it all together\n",
    "\n",
    "Let's try using our new function in a for loop. This for loop is a bit different from the previous ones; it's actually simpler. Instead of using the PSSM, we can simply iterate over the positions in the human sequence, get the equivalent amino acid in the cat sequence, and use our function to ask whether the amino acids at that position have different hydrophobicity.\n",
    "\n",
    "One new addition to this algorithm (besides our custom function) is the [`range()` function](https://github.com/wilfredinni/python-cheatsheet#for-loops-and-the-range-function)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 1,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 7,\n",
       " 8,\n",
       " 9,\n",
       " 10,\n",
       " 11,\n",
       " 12,\n",
       " 13,\n",
       " 14,\n",
       " 15,\n",
       " 16,\n",
       " 17,\n",
       " 18,\n",
       " 19,\n",
       " 20,\n",
       " 21,\n",
       " 22,\n",
       " 23,\n",
       " 24,\n",
       " 25,\n",
       " 26,\n",
       " 27,\n",
       " 28,\n",
       " 29,\n",
       " 30,\n",
       " 31,\n",
       " 32,\n",
       " 33,\n",
       " 34,\n",
       " 35,\n",
       " 36,\n",
       " 37,\n",
       " 38,\n",
       " 39,\n",
       " 40,\n",
       " 41,\n",
       " 42,\n",
       " 43,\n",
       " 44,\n",
       " 45,\n",
       " 46,\n",
       " 47,\n",
       " 48,\n",
       " 49,\n",
       " 50,\n",
       " 51,\n",
       " 52,\n",
       " 53,\n",
       " 54,\n",
       " 55,\n",
       " 56,\n",
       " 57,\n",
       " 58,\n",
       " 59,\n",
       " 60,\n",
       " 61,\n",
       " 62,\n",
       " 63,\n",
       " 64,\n",
       " 65,\n",
       " 66,\n",
       " 67,\n",
       " 68,\n",
       " 69,\n",
       " 70,\n",
       " 71,\n",
       " 72,\n",
       " 73,\n",
       " 74,\n",
       " 75,\n",
       " 76,\n",
       " 77,\n",
       " 78,\n",
       " 79,\n",
       " 80,\n",
       " 81,\n",
       " 82,\n",
       " 83,\n",
       " 84,\n",
       " 85,\n",
       " 86,\n",
       " 87,\n",
       " 88,\n",
       " 89,\n",
       " 90,\n",
       " 91,\n",
       " 92,\n",
       " 93,\n",
       " 94,\n",
       " 95,\n",
       " 96,\n",
       " 97,\n",
       " 98,\n",
       " 99,\n",
       " 100,\n",
       " 101,\n",
       " 102,\n",
       " 103,\n",
       " 104,\n",
       " 105,\n",
       " 106,\n",
       " 107,\n",
       " 108,\n",
       " 109,\n",
       " 110,\n",
       " 111,\n",
       " 112,\n",
       " 113,\n",
       " 114,\n",
       " 115,\n",
       " 116,\n",
       " 117,\n",
       " 118,\n",
       " 119,\n",
       " 120,\n",
       " 121,\n",
       " 122,\n",
       " 123,\n",
       " 124,\n",
       " 125,\n",
       " 126,\n",
       " 127,\n",
       " 128,\n",
       " 129,\n",
       " 130,\n",
       " 131,\n",
       " 132,\n",
       " 133,\n",
       " 134,\n",
       " 135,\n",
       " 136,\n",
       " 137,\n",
       " 138,\n",
       " 139,\n",
       " 140,\n",
       " 141,\n",
       " 142,\n",
       " 143,\n",
       " 144,\n",
       " 145,\n",
       " 146,\n",
       " 147,\n",
       " 148,\n",
       " 149,\n",
       " 150,\n",
       " 151,\n",
       " 152,\n",
       " 153,\n",
       " 154,\n",
       " 155,\n",
       " 156,\n",
       " 157,\n",
       " 158,\n",
       " 159,\n",
       " 160,\n",
       " 161,\n",
       " 162,\n",
       " 163,\n",
       " 164,\n",
       " 165,\n",
       " 166,\n",
       " 167,\n",
       " 168,\n",
       " 169,\n",
       " 170,\n",
       " 171,\n",
       " 172,\n",
       " 173,\n",
       " 174,\n",
       " 175,\n",
       " 176,\n",
       " 177,\n",
       " 178,\n",
       " 179,\n",
       " 180,\n",
       " 181,\n",
       " 182,\n",
       " 183,\n",
       " 184,\n",
       " 185,\n",
       " 186,\n",
       " 187,\n",
       " 188,\n",
       " 189,\n",
       " 190,\n",
       " 191,\n",
       " 192,\n",
       " 193,\n",
       " 194,\n",
       " 195,\n",
       " 196,\n",
       " 197,\n",
       " 198,\n",
       " 199,\n",
       " 200,\n",
       " 201,\n",
       " 202,\n",
       " 203,\n",
       " 204,\n",
       " 205,\n",
       " 206,\n",
       " 207,\n",
       " 208,\n",
       " 209,\n",
       " 210,\n",
       " 211,\n",
       " 212,\n",
       " 213,\n",
       " 214,\n",
       " 215,\n",
       " 216,\n",
       " 217,\n",
       " 218,\n",
       " 219,\n",
       " 220,\n",
       " 221,\n",
       " 222,\n",
       " 223,\n",
       " 224,\n",
       " 225,\n",
       " 226,\n",
       " 227,\n",
       " 228,\n",
       " 229,\n",
       " 230,\n",
       " 231,\n",
       " 232,\n",
       " 233,\n",
       " 234,\n",
       " 235,\n",
       " 236,\n",
       " 237,\n",
       " 238,\n",
       " 239,\n",
       " 240,\n",
       " 241,\n",
       " 242,\n",
       " 243,\n",
       " 244,\n",
       " 245,\n",
       " 246,\n",
       " 247,\n",
       " 248,\n",
       " 249,\n",
       " 250,\n",
       " 251,\n",
       " 252,\n",
       " 253,\n",
       " 254,\n",
       " 255,\n",
       " 256,\n",
       " 257,\n",
       " 258,\n",
       " 259,\n",
       " 260,\n",
       " 261,\n",
       " 262,\n",
       " 263,\n",
       " 264,\n",
       " 265,\n",
       " 266,\n",
       " 267,\n",
       " 268,\n",
       " 269,\n",
       " 270,\n",
       " 271,\n",
       " 272,\n",
       " 273,\n",
       " 274,\n",
       " 275,\n",
       " 276,\n",
       " 277,\n",
       " 278,\n",
       " 279,\n",
       " 280,\n",
       " 281,\n",
       " 282,\n",
       " 283,\n",
       " 284,\n",
       " 285,\n",
       " 286,\n",
       " 287,\n",
       " 288,\n",
       " 289,\n",
       " 290,\n",
       " 291,\n",
       " 292,\n",
       " 293,\n",
       " 294,\n",
       " 295,\n",
       " 296,\n",
       " 297,\n",
       " 298,\n",
       " 299,\n",
       " 300,\n",
       " 301,\n",
       " 302,\n",
       " 303,\n",
       " 304,\n",
       " 305,\n",
       " 306,\n",
       " 307,\n",
       " 308,\n",
       " 309,\n",
       " 310,\n",
       " 311,\n",
       " 312,\n",
       " 313,\n",
       " 314,\n",
       " 315,\n",
       " 316,\n",
       " 317,\n",
       " 318,\n",
       " 319,\n",
       " 320,\n",
       " 321,\n",
       " 322,\n",
       " 323,\n",
       " 324,\n",
       " 325,\n",
       " 326,\n",
       " 327,\n",
       " 328,\n",
       " 329,\n",
       " 330,\n",
       " 331,\n",
       " 332,\n",
       " 333,\n",
       " 334,\n",
       " 335,\n",
       " 336,\n",
       " 337,\n",
       " 338,\n",
       " 339,\n",
       " 340,\n",
       " 341,\n",
       " 342,\n",
       " 343,\n",
       " 344,\n",
       " 345,\n",
       " 346,\n",
       " 347,\n",
       " 348,\n",
       " 349,\n",
       " 350,\n",
       " 351,\n",
       " 352,\n",
       " 353,\n",
       " 354,\n",
       " 355,\n",
       " 356,\n",
       " 357,\n",
       " 358,\n",
       " 359,\n",
       " 360,\n",
       " 361,\n",
       " 362,\n",
       " 363,\n",
       " 364,\n",
       " 365,\n",
       " 366,\n",
       " 367,\n",
       " 368,\n",
       " 369,\n",
       " 370,\n",
       " 371,\n",
       " 372,\n",
       " 373,\n",
       " 374,\n",
       " 375,\n",
       " 376,\n",
       " 377,\n",
       " 378,\n",
       " 379,\n",
       " 380,\n",
       " 381,\n",
       " 382,\n",
       " 383,\n",
       " 384,\n",
       " 385,\n",
       " 386,\n",
       " 387,\n",
       " 388,\n",
       " 389,\n",
       " 390,\n",
       " 391,\n",
       " 392,\n",
       " 393,\n",
       " 394,\n",
       " 395,\n",
       " 396,\n",
       " 397,\n",
       " 398,\n",
       " 399,\n",
       " 400,\n",
       " 401,\n",
       " 402,\n",
       " 403,\n",
       " 404,\n",
       " 405,\n",
       " 406,\n",
       " 407,\n",
       " 408,\n",
       " 409,\n",
       " 410,\n",
       " 411,\n",
       " 412,\n",
       " 413,\n",
       " 414,\n",
       " 415,\n",
       " 416,\n",
       " 417,\n",
       " 418,\n",
       " 419,\n",
       " 420,\n",
       " 421,\n",
       " 422,\n",
       " 423,\n",
       " 424,\n",
       " 425,\n",
       " 426,\n",
       " 427,\n",
       " 428,\n",
       " 429,\n",
       " 430,\n",
       " 431,\n",
       " 432,\n",
       " 433,\n",
       " 434,\n",
       " 435,\n",
       " 436,\n",
       " 437,\n",
       " 438,\n",
       " 439,\n",
       " 440,\n",
       " 441,\n",
       " 442,\n",
       " 443,\n",
       " 444,\n",
       " 445,\n",
       " 446,\n",
       " 447,\n",
       " 448,\n",
       " 449,\n",
       " 450,\n",
       " 451,\n",
       " 452,\n",
       " 453,\n",
       " 454,\n",
       " 455,\n",
       " 456,\n",
       " 457,\n",
       " 458,\n",
       " 459,\n",
       " 460,\n",
       " 461,\n",
       " 462,\n",
       " 463,\n",
       " 464,\n",
       " 465,\n",
       " 466,\n",
       " 467,\n",
       " 468,\n",
       " 469,\n",
       " 470,\n",
       " 471,\n",
       " 472,\n",
       " 473,\n",
       " 474,\n",
       " 475,\n",
       " 476,\n",
       " 477,\n",
       " 478,\n",
       " 479,\n",
       " 480,\n",
       " 481,\n",
       " 482,\n",
       " 483,\n",
       " 484,\n",
       " 485,\n",
       " 486,\n",
       " 487,\n",
       " 488,\n",
       " 489,\n",
       " 490,\n",
       " 491,\n",
       " 492,\n",
       " 493,\n",
       " 494,\n",
       " 495,\n",
       " 496,\n",
       " 497,\n",
       " 498,\n",
       " 499,\n",
       " 500,\n",
       " 501,\n",
       " 502,\n",
       " 503,\n",
       " 504,\n",
       " 505,\n",
       " 506,\n",
       " 507,\n",
       " 508,\n",
       " 509,\n",
       " 510,\n",
       " 511,\n",
       " 512,\n",
       " 513,\n",
       " 514,\n",
       " 515,\n",
       " 516,\n",
       " 517,\n",
       " 518,\n",
       " 519,\n",
       " 520,\n",
       " 521,\n",
       " 522,\n",
       " 523,\n",
       " 524,\n",
       " 525,\n",
       " 526,\n",
       " 527,\n",
       " 528,\n",
       " 529,\n",
       " 530,\n",
       " 531,\n",
       " 532,\n",
       " 533,\n",
       " 534,\n",
       " 535,\n",
       " 536,\n",
       " 537,\n",
       " 538,\n",
       " 539,\n",
       " 540,\n",
       " 541,\n",
       " 542,\n",
       " 543,\n",
       " 544,\n",
       " 545,\n",
       " 546,\n",
       " 547,\n",
       " 548,\n",
       " 549,\n",
       " 550,\n",
       " 551,\n",
       " 552,\n",
       " 553,\n",
       " 554,\n",
       " 555,\n",
       " 556,\n",
       " 557,\n",
       " 558,\n",
       " 559,\n",
       " 560,\n",
       " 561,\n",
       " 562,\n",
       " 563,\n",
       " 564,\n",
       " 565,\n",
       " 566,\n",
       " 567,\n",
       " 568,\n",
       " 569,\n",
       " 570,\n",
       " 571,\n",
       " 572,\n",
       " 573,\n",
       " 574,\n",
       " 575,\n",
       " 576,\n",
       " 577,\n",
       " 578,\n",
       " 579,\n",
       " 580,\n",
       " 581,\n",
       " 582,\n",
       " 583,\n",
       " 584,\n",
       " 585,\n",
       " 586,\n",
       " 587,\n",
       " 588,\n",
       " 589,\n",
       " 590,\n",
       " 591,\n",
       " 592,\n",
       " 593,\n",
       " 594,\n",
       " 595,\n",
       " 596,\n",
       " 597,\n",
       " 598,\n",
       " 599,\n",
       " 600,\n",
       " 601,\n",
       " 602,\n",
       " 603,\n",
       " 604,\n",
       " 605,\n",
       " 606,\n",
       " 607,\n",
       " 608,\n",
       " 609,\n",
       " 610,\n",
       " 611,\n",
       " 612,\n",
       " 613,\n",
       " 614,\n",
       " 615,\n",
       " 616,\n",
       " 617,\n",
       " 618,\n",
       " 619,\n",
       " 620,\n",
       " 621,\n",
       " 622,\n",
       " 623,\n",
       " 624,\n",
       " 625,\n",
       " 626,\n",
       " 627,\n",
       " 628,\n",
       " 629,\n",
       " 630,\n",
       " 631,\n",
       " 632,\n",
       " 633,\n",
       " 634,\n",
       " 635,\n",
       " 636,\n",
       " 637,\n",
       " 638,\n",
       " 639,\n",
       " 640,\n",
       " 641,\n",
       " 642,\n",
       " 643,\n",
       " 644,\n",
       " 645,\n",
       " 646,\n",
       " 647,\n",
       " 648,\n",
       " 649,\n",
       " 650,\n",
       " 651,\n",
       " 652,\n",
       " 653,\n",
       " 654,\n",
       " 655,\n",
       " 656,\n",
       " 657,\n",
       " 658,\n",
       " 659,\n",
       " 660,\n",
       " 661,\n",
       " 662,\n",
       " 663,\n",
       " 664,\n",
       " 665,\n",
       " 666,\n",
       " 667,\n",
       " 668,\n",
       " 669,\n",
       " 670,\n",
       " 671,\n",
       " 672,\n",
       " 673,\n",
       " 674,\n",
       " 675,\n",
       " 676,\n",
       " 677,\n",
       " 678,\n",
       " 679,\n",
       " 680,\n",
       " 681,\n",
       " 682,\n",
       " 683,\n",
       " 684,\n",
       " 685,\n",
       " 686,\n",
       " 687,\n",
       " 688,\n",
       " 689,\n",
       " 690,\n",
       " 691,\n",
       " 692,\n",
       " 693,\n",
       " 694,\n",
       " 695,\n",
       " 696,\n",
       " 697,\n",
       " 698,\n",
       " 699,\n",
       " 700,\n",
       " 701,\n",
       " 702,\n",
       " 703,\n",
       " 704,\n",
       " 705,\n",
       " 706,\n",
       " 707,\n",
       " 708,\n",
       " 709,\n",
       " 710,\n",
       " 711,\n",
       " 712,\n",
       " 713,\n",
       " 714,\n",
       " 715,\n",
       " 716,\n",
       " 717,\n",
       " 718,\n",
       " 719,\n",
       " 720,\n",
       " 721,\n",
       " 722,\n",
       " 723,\n",
       " 724,\n",
       " 725,\n",
       " 726,\n",
       " 727,\n",
       " 728,\n",
       " 729,\n",
       " 730,\n",
       " 731,\n",
       " 732,\n",
       " 733,\n",
       " 734,\n",
       " 735,\n",
       " 736,\n",
       " 737,\n",
       " 738,\n",
       " 739,\n",
       " 740,\n",
       " 741,\n",
       " 742,\n",
       " 743,\n",
       " 744,\n",
       " 745,\n",
       " 746,\n",
       " 747,\n",
       " 748,\n",
       " 749,\n",
       " 750,\n",
       " 751,\n",
       " 752,\n",
       " 753,\n",
       " 754,\n",
       " 755,\n",
       " 756,\n",
       " 757,\n",
       " 758,\n",
       " 759,\n",
       " 760,\n",
       " 761,\n",
       " 762,\n",
       " 763,\n",
       " 764,\n",
       " 765,\n",
       " 766,\n",
       " 767,\n",
       " 768,\n",
       " 769,\n",
       " 770,\n",
       " 771,\n",
       " 772,\n",
       " 773,\n",
       " 774,\n",
       " 775,\n",
       " 776,\n",
       " 777,\n",
       " 778,\n",
       " 779,\n",
       " 780,\n",
       " 781,\n",
       " 782,\n",
       " 783,\n",
       " 784,\n",
       " 785,\n",
       " 786,\n",
       " 787,\n",
       " 788,\n",
       " 789,\n",
       " 790,\n",
       " 791,\n",
       " 792,\n",
       " 793,\n",
       " 794,\n",
       " 795,\n",
       " 796,\n",
       " 797,\n",
       " 798,\n",
       " 799,\n",
       " 800,\n",
       " 801,\n",
       " 802,\n",
       " 803,\n",
       " 804,\n",
       " 805,\n",
       " 806,\n",
       " 807,\n",
       " 808,\n",
       " 809,\n",
       " 810,\n",
       " 811,\n",
       " 812,\n",
       " 813,\n",
       " 814,\n",
       " 815,\n",
       " 816,\n",
       " 817,\n",
       " 818,\n",
       " 819,\n",
       " 820,\n",
       " 821,\n",
       " 822,\n",
       " 823,\n",
       " 824,\n",
       " 825,\n",
       " 826,\n",
       " 827,\n",
       " 828,\n",
       " 829,\n",
       " 830,\n",
       " 831,\n",
       " 832,\n",
       " 833,\n",
       " 834,\n",
       " 835,\n",
       " 836,\n",
       " 837,\n",
       " 838,\n",
       " 839,\n",
       " 840,\n",
       " 841,\n",
       " 842,\n",
       " 843,\n",
       " 844,\n",
       " 845,\n",
       " 846,\n",
       " 847,\n",
       " 848,\n",
       " 849,\n",
       " 850,\n",
       " 851,\n",
       " 852,\n",
       " 853,\n",
       " 854,\n",
       " 855,\n",
       " 856,\n",
       " 857,\n",
       " 858,\n",
       " 859,\n",
       " 860,\n",
       " 861,\n",
       " 862,\n",
       " 863,\n",
       " 864,\n",
       " 865,\n",
       " 866,\n",
       " 867,\n",
       " 868,\n",
       " 869,\n",
       " 870,\n",
       " 871,\n",
       " 872,\n",
       " 873,\n",
       " 874,\n",
       " 875,\n",
       " 876,\n",
       " 877,\n",
       " 878,\n",
       " 879,\n",
       " 880,\n",
       " 881,\n",
       " 882,\n",
       " 883,\n",
       " 884,\n",
       " 885,\n",
       " 886,\n",
       " 887,\n",
       " 888,\n",
       " 889,\n",
       " 890,\n",
       " 891,\n",
       " 892,\n",
       " 893,\n",
       " 894,\n",
       " 895,\n",
       " 896,\n",
       " 897,\n",
       " 898,\n",
       " 899,\n",
       " 900,\n",
       " 901,\n",
       " 902,\n",
       " 903,\n",
       " 904,\n",
       " 905,\n",
       " 906,\n",
       " 907,\n",
       " 908,\n",
       " 909,\n",
       " 910,\n",
       " 911,\n",
       " 912,\n",
       " 913,\n",
       " 914,\n",
       " 915,\n",
       " 916,\n",
       " 917,\n",
       " 918,\n",
       " 919,\n",
       " 920,\n",
       " 921,\n",
       " 922,\n",
       " 923,\n",
       " 924,\n",
       " 925,\n",
       " 926,\n",
       " 927,\n",
       " 928,\n",
       " 929,\n",
       " 930,\n",
       " 931,\n",
       " 932,\n",
       " 933,\n",
       " 934,\n",
       " 935,\n",
       " 936,\n",
       " 937,\n",
       " 938,\n",
       " 939,\n",
       " 940,\n",
       " 941,\n",
       " 942,\n",
       " 943,\n",
       " 944,\n",
       " 945,\n",
       " 946,\n",
       " 947,\n",
       " 948,\n",
       " 949,\n",
       " 950,\n",
       " 951,\n",
       " 952,\n",
       " 953,\n",
       " 954,\n",
       " 955,\n",
       " 956,\n",
       " 957,\n",
       " 958,\n",
       " 959,\n",
       " 960,\n",
       " 961,\n",
       " 962,\n",
       " 963,\n",
       " 964,\n",
       " 965,\n",
       " 966,\n",
       " 967,\n",
       " 968,\n",
       " 969,\n",
       " 970,\n",
       " 971,\n",
       " 972,\n",
       " 973,\n",
       " 974,\n",
       " 975,\n",
       " 976,\n",
       " 977,\n",
       " 978,\n",
       " 979,\n",
       " 980,\n",
       " 981,\n",
       " 982,\n",
       " 983,\n",
       " 984,\n",
       " 985,\n",
       " 986,\n",
       " 987,\n",
       " 988,\n",
       " 989,\n",
       " 990,\n",
       " 991,\n",
       " 992,\n",
       " 993,\n",
       " 994,\n",
       " 995,\n",
       " 996,\n",
       " 997,\n",
       " 998,\n",
       " 999,\n",
       " ...]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(range(len(human_compact)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "comparing hydrophobicity between M and L\n",
      "the difference is hydrophobicity is 2.0\n",
      "comparing hydrophobicity between P and S\n",
      "the difference is hydrophobicity is 1.0\n",
      "comparing hydrophobicity between P and R\n",
      "the difference is hydrophobicity is 2.0\n",
      "comparing hydrophobicity between A and M\n",
      "the difference is hydrophobicity is 4.0\n",
      "comparing hydrophobicity between P and P\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 6\n",
      "detected a deletion at position 7\n",
      "detected a deletion at position 8\n",
      "detected a deletion at position 9\n",
      "detected a deletion at position 10\n",
      "detected a deletion at position 11\n",
      "detected a deletion at position 12\n",
      "detected a deletion at position 13\n",
      "detected a deletion at position 14\n",
      "detected a deletion at position 15\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 17\n",
      "detected a deletion at position 18\n",
      "detected a deletion at position 19\n",
      "comparing hydrophobicity between H and V\n",
      "the difference is hydrophobicity is 6.0\n",
      "detected a change in hydrophobicity at position 20\n",
      "comparing hydrophobicity between I and P\n",
      "the difference is hydrophobicity is 12.0\n",
      "detected a change in hydrophobicity at position 21\n",
      "comparing hydrophobicity between E and S\n",
      "the difference is hydrophobicity is 4.0\n",
      "comparing hydrophobicity between H and L\n",
      "the difference is hydrophobicity is 7.0\n",
      "detected a change in hydrophobicity at position 23\n",
      "comparing hydrophobicity between S and S\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 25\n",
      "detected a deletion at position 26\n",
      "detected a deletion at position 27\n",
      "detected a deletion at position 28\n",
      "detected a deletion at position 29\n",
      "comparing hydrophobicity between L and P\n",
      "the difference is hydrophobicity is 10.0\n",
      "detected a change in hydrophobicity at position 30\n",
      "detected a deletion at position 31\n",
      "detected a deletion at position 32\n",
      "detected a deletion at position 33\n",
      "detected a deletion at position 34\n",
      "detected a deletion at position 35\n",
      "detected a deletion at position 36\n",
      "detected a deletion at position 37\n",
      "detected a deletion at position 38\n",
      "detected a deletion at position 39\n",
      "detected a deletion at position 40\n",
      "detected a deletion at position 41\n",
      "comparing hydrophobicity between S and C\n",
      "the difference is hydrophobicity is 7.0\n",
      "detected a change in hydrophobicity at position 42\n",
      "detected a deletion at position 43\n",
      "detected a deletion at position 44\n",
      "detected a deletion at position 45\n",
      "comparing hydrophobicity between P and T\n",
      "the difference is hydrophobicity is 1.0\n",
      "comparing hydrophobicity between A and A\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between I and A\n",
      "the difference is hydrophobicity is 8.0\n",
      "detected a change in hydrophobicity at position 48\n",
      "comparing hydrophobicity between G and H\n",
      "the difference is hydrophobicity is 1.0\n",
      "comparing hydrophobicity between P and C\n",
      "the difference is hydrophobicity is 6.0\n",
      "detected a change in hydrophobicity at position 50\n",
      "detected a deletion at position 51\n",
      "detected a deletion at position 52\n",
      "detected a deletion at position 53\n",
      "detected a deletion at position 54\n",
      "detected a deletion at position 55\n",
      "detected a deletion at position 56\n",
      "detected a deletion at position 57\n",
      "detected a deletion at position 58\n",
      "detected a deletion at position 59\n",
      "detected a deletion at position 60\n",
      "detected a deletion at position 61\n",
      "comparing hydrophobicity between P and V\n",
      "the difference is hydrophobicity is 9.0\n",
      "detected a change in hydrophobicity at position 62\n",
      "detected a deletion at position 63\n",
      "detected a deletion at position 64\n",
      "detected a deletion at position 65\n",
      "detected a deletion at position 66\n",
      "detected a deletion at position 67\n",
      "detected a deletion at position 68\n",
      "detected a deletion at position 69\n",
      "detected a deletion at position 70\n",
      "detected a deletion at position 71\n",
      "detected a deletion at position 72\n",
      "detected a deletion at position 73\n",
      "detected a deletion at position 74\n",
      "detected a deletion at position 75\n",
      "comparing hydrophobicity between V and L\n",
      "the difference is hydrophobicity is 1.0\n",
      "detected a deletion at position 77\n",
      "detected a deletion at position 78\n",
      "detected a deletion at position 79\n",
      "detected a deletion at position 80\n",
      "detected a deletion at position 81\n",
      "detected a deletion at position 82\n",
      "detected a deletion at position 83\n",
      "detected a deletion at position 84\n",
      "detected a deletion at position 85\n",
      "detected a deletion at position 86\n",
      "detected a deletion at position 87\n",
      "detected a deletion at position 88\n",
      "detected a deletion at position 89\n",
      "detected a deletion at position 90\n",
      "detected a deletion at position 91\n",
      "detected a deletion at position 92\n",
      "detected a deletion at position 93\n",
      "detected a deletion at position 94\n",
      "detected a deletion at position 95\n",
      "detected a deletion at position 96\n",
      "detected a deletion at position 97\n",
      "detected a deletion at position 98\n",
      "comparing hydrophobicity between N and L\n",
      "the difference is hydrophobicity is 13.0\n",
      "detected a change in hydrophobicity at position 99\n",
      "detected a deletion at position 100\n",
      "detected a deletion at position 101\n",
      "detected a deletion at position 102\n",
      "detected a deletion at position 103\n",
      "detected a deletion at position 104\n",
      "detected a deletion at position 105\n",
      "comparing hydrophobicity between P and P\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 107\n",
      "detected a deletion at position 108\n",
      "detected a deletion at position 109\n",
      "comparing hydrophobicity between S and G\n",
      "the difference is hydrophobicity is 3.0\n",
      "detected a deletion at position 111\n",
      "detected a deletion at position 112\n",
      "comparing hydrophobicity between V and C\n",
      "the difference is hydrophobicity is 3.0\n",
      "detected a deletion at position 114\n",
      "detected a deletion at position 115\n",
      "detected a deletion at position 116\n",
      "comparing hydrophobicity between K and G\n",
      "the difference is hydrophobicity is 9.0\n",
      "detected a change in hydrophobicity at position 117\n",
      "comparing hydrophobicity between T and W\n",
      "the difference is hydrophobicity is 6.0\n",
      "detected a change in hydrophobicity at position 118\n",
      "comparing hydrophobicity between K and G\n",
      "the difference is hydrophobicity is 9.0\n",
      "detected a change in hydrophobicity at position 119\n",
      "detected a deletion at position 120\n",
      "detected a deletion at position 121\n",
      "detected a deletion at position 122\n",
      "detected a deletion at position 123\n",
      "detected a deletion at position 124\n",
      "detected a deletion at position 125\n",
      "detected a deletion at position 126\n",
      "detected a deletion at position 127\n",
      "detected a deletion at position 128\n",
      "detected a deletion at position 129\n",
      "detected a deletion at position 130\n",
      "detected a deletion at position 131\n",
      "detected a deletion at position 132\n",
      "detected a deletion at position 133\n",
      "detected a deletion at position 134\n",
      "detected a deletion at position 135\n",
      "detected a deletion at position 136\n",
      "detected a deletion at position 137\n",
      "detected a deletion at position 138\n",
      "detected a deletion at position 139\n",
      "detected a deletion at position 140\n",
      "detected a deletion at position 141\n",
      "detected a deletion at position 142\n",
      "detected a deletion at position 143\n",
      "detected a deletion at position 144\n",
      "detected a deletion at position 145\n",
      "detected a deletion at position 146\n",
      "detected a deletion at position 147\n",
      "detected a deletion at position 148\n",
      "detected a deletion at position 149\n",
      "detected a deletion at position 150\n",
      "detected a deletion at position 151\n",
      "comparing hydrophobicity between N and Q\n",
      "the difference is hydrophobicity is 1.0\n",
      "comparing hydrophobicity between S and P\n",
      "the difference is hydrophobicity is 1.0\n",
      "comparing hydrophobicity between G and S\n",
      "the difference is hydrophobicity is 3.0\n",
      "comparing hydrophobicity between I and L\n",
      "the difference is hydrophobicity is 2.0\n",
      "detected a deletion at position 156\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 158\n",
      "detected a deletion at position 159\n",
      "detected a deletion at position 160\n",
      "detected a deletion at position 161\n",
      "comparing hydrophobicity between T and P\n",
      "the difference is hydrophobicity is 1.0\n",
      "detected a deletion at position 163\n",
      "detected a deletion at position 164\n",
      "detected a deletion at position 165\n",
      "detected a deletion at position 166\n",
      "detected a deletion at position 167\n",
      "detected a deletion at position 168\n",
      "detected a deletion at position 169\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 171\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 173\n",
      "detected a deletion at position 174\n",
      "detected a deletion at position 175\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 177\n",
      "detected a deletion at position 178\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 180\n",
      "comparing hydrophobicity between D and S\n",
      "the difference is hydrophobicity is 5.0\n",
      "comparing hydrophobicity between E and G\n",
      "the difference is hydrophobicity is 7.0\n",
      "detected a change in hydrophobicity at position 182\n",
      "detected a deletion at position 183\n",
      "comparing hydrophobicity between R and P\n",
      "the difference is hydrophobicity is 2.0\n",
      "comparing hydrophobicity between C and L\n",
      "the difference is hydrophobicity is 4.0\n",
      "detected a deletion at position 186\n",
      "detected a deletion at position 187\n",
      "detected a deletion at position 188\n",
      "detected a deletion at position 189\n",
      "detected a deletion at position 190\n",
      "detected a deletion at position 191\n",
      "detected a deletion at position 192\n",
      "detected a deletion at position 193\n",
      "detected a deletion at position 194\n",
      "detected a deletion at position 195\n",
      "detected a deletion at position 196\n",
      "detected a deletion at position 197\n",
      "detected a deletion at position 198\n",
      "detected a deletion at position 199\n",
      "detected a deletion at position 200\n",
      "detected a deletion at position 201\n",
      "detected a deletion at position 202\n",
      "detected a deletion at position 203\n",
      "detected a deletion at position 204\n",
      "detected a deletion at position 205\n",
      "detected a deletion at position 206\n",
      "detected a deletion at position 207\n",
      "detected a deletion at position 208\n",
      "detected a deletion at position 209\n",
      "detected a deletion at position 210\n",
      "detected a deletion at position 211\n",
      "detected a deletion at position 212\n",
      "detected a deletion at position 213\n",
      "detected a deletion at position 214\n",
      "detected a deletion at position 215\n",
      "detected a deletion at position 216\n",
      "detected a deletion at position 217\n",
      "detected a deletion at position 218\n",
      "detected a deletion at position 219\n",
      "detected a deletion at position 220\n",
      "detected a deletion at position 221\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 223\n",
      "detected a deletion at position 224\n",
      "detected a deletion at position 225\n",
      "detected a deletion at position 226\n",
      "detected a deletion at position 227\n",
      "comparing hydrophobicity between K and L\n",
      "the difference is hydrophobicity is 17.0\n",
      "detected a change in hydrophobicity at position 228\n",
      "detected a deletion at position 229\n",
      "detected a deletion at position 230\n",
      "detected a deletion at position 231\n",
      "detected a deletion at position 232\n",
      "detected a deletion at position 233\n",
      "detected a deletion at position 234\n",
      "detected a deletion at position 235\n",
      "detected a deletion at position 236\n",
      "detected a deletion at position 237\n",
      "detected a deletion at position 238\n",
      "detected a deletion at position 239\n",
      "detected a deletion at position 240\n",
      "detected a deletion at position 241\n",
      "detected a deletion at position 242\n",
      "detected a deletion at position 243\n",
      "detected a deletion at position 244\n",
      "detected a deletion at position 245\n",
      "detected a deletion at position 246\n",
      "detected a deletion at position 247\n",
      "detected a deletion at position 248\n",
      "detected a deletion at position 249\n",
      "detected a deletion at position 250\n",
      "detected a deletion at position 251\n",
      "detected a deletion at position 252\n",
      "detected a deletion at position 253\n",
      "detected a deletion at position 254\n",
      "detected a deletion at position 255\n",
      "detected a deletion at position 256\n",
      "detected a deletion at position 257\n",
      "comparing hydrophobicity between I and L\n",
      "the difference is hydrophobicity is 2.0\n",
      "detected a deletion at position 259\n",
      "* position K259 is a residue of interest!\n",
      "detected a deletion at position 260\n",
      "* position K260 is a residue of interest!\n",
      "detected a deletion at position 261\n",
      "detected a deletion at position 262\n",
      "detected a deletion at position 263\n",
      "detected a deletion at position 264\n",
      "detected a deletion at position 265\n",
      "detected a deletion at position 266\n",
      "detected a deletion at position 267\n",
      "detected a deletion at position 268\n",
      "detected a deletion at position 269\n",
      "detected a deletion at position 270\n",
      "detected a deletion at position 271\n",
      "detected a deletion at position 272\n",
      "detected a deletion at position 273\n",
      "detected a deletion at position 274\n",
      "detected a deletion at position 275\n",
      "detected a deletion at position 276\n",
      "detected a deletion at position 277\n",
      "detected a deletion at position 278\n",
      "detected a deletion at position 279\n",
      "detected a deletion at position 280\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 283\n",
      "detected a deletion at position 284\n",
      "detected a deletion at position 285\n",
      "detected a deletion at position 286\n",
      "detected a deletion at position 287\n",
      "detected a deletion at position 288\n",
      "detected a deletion at position 289\n",
      "detected a deletion at position 290\n",
      "detected a deletion at position 291\n",
      "comparing hydrophobicity between R and R\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between I and I\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 294\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 297\n",
      "detected a deletion at position 298\n",
      "detected a deletion at position 299\n",
      "detected a deletion at position 300\n",
      "detected a deletion at position 301\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 303\n",
      "detected a deletion at position 304\n",
      "comparing hydrophobicity between P and P\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between W and W\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between Q and Q\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 308\n",
      "detected a deletion at position 309\n",
      "comparing hydrophobicity between L and L\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 311\n",
      "detected a deletion at position 312\n",
      "detected a deletion at position 313\n",
      "detected a deletion at position 314\n",
      "detected a deletion at position 315\n",
      "comparing hydrophobicity between H and H\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 317\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 320\n",
      "detected a deletion at position 321\n",
      "detected a deletion at position 322\n",
      "detected a deletion at position 323\n",
      "detected a deletion at position 324\n",
      "detected a deletion at position 325\n",
      "detected a deletion at position 326\n",
      "comparing hydrophobicity between W and W\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 328\n",
      "detected a deletion at position 329\n",
      "detected a deletion at position 330\n",
      "comparing hydrophobicity between A and A\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between A and A\n",
      "the difference is hydrophobicity is 0.0\n",
      "* position H332 is a residue of interest!\n",
      "comparing hydrophobicity between H and H\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 335\n",
      "detected a deletion at position 336\n",
      "detected a deletion at position 337\n",
      "detected a deletion at position 338\n",
      "detected a deletion at position 339\n",
      "detected a deletion at position 340\n",
      "detected a deletion at position 341\n",
      "detected a deletion at position 342\n",
      "detected a deletion at position 343\n",
      "detected a deletion at position 344\n",
      "detected a deletion at position 345\n",
      "detected a deletion at position 346\n",
      "detected a deletion at position 347\n",
      "detected a deletion at position 348\n",
      "detected a deletion at position 349\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 351\n",
      "detected a deletion at position 352\n",
      "detected a deletion at position 353\n",
      "detected a deletion at position 354\n",
      "detected a deletion at position 355\n",
      "detected a deletion at position 356\n",
      "detected a deletion at position 357\n",
      "detected a deletion at position 358\n",
      "detected a deletion at position 359\n",
      "detected a deletion at position 360\n",
      "detected a deletion at position 361\n",
      "detected a deletion at position 362\n",
      "detected a deletion at position 363\n",
      "detected a deletion at position 364\n",
      "detected a deletion at position 365\n",
      "detected a deletion at position 366\n",
      "detected a deletion at position 367\n",
      "detected a deletion at position 368\n",
      "detected a deletion at position 369\n",
      "detected a deletion at position 370\n",
      "comparing hydrophobicity between H and H\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 372\n",
      "detected a deletion at position 373\n",
      "detected a deletion at position 374\n",
      "detected a deletion at position 375\n",
      "detected a deletion at position 376\n",
      "detected a deletion at position 377\n",
      "detected a deletion at position 378\n",
      "detected a deletion at position 379\n",
      "detected a deletion at position 380\n",
      "detected a deletion at position 381\n",
      "* position D381 is a residue of interest!\n",
      "comparing hydrophobicity between D and D\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between I and I\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between A and A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between L and L\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 386\n",
      "detected a deletion at position 387\n",
      "comparing hydrophobicity between L and L\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 389\n",
      "detected a deletion at position 390\n",
      "detected a deletion at position 391\n",
      "detected a deletion at position 392\n",
      "detected a deletion at position 393\n",
      "detected a deletion at position 394\n",
      "detected a deletion at position 395\n",
      "detected a deletion at position 396\n",
      "detected a deletion at position 397\n",
      "detected a deletion at position 398\n",
      "detected a deletion at position 399\n",
      "comparing hydrophobicity between P and P\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 401\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between L and L\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between P and P\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 405\n",
      "detected a deletion at position 406\n",
      "detected a deletion at position 407\n",
      "detected a deletion at position 408\n",
      "detected a deletion at position 409\n",
      "detected a deletion at position 410\n",
      "detected a deletion at position 411\n",
      "detected a deletion at position 412\n",
      "detected a deletion at position 413\n",
      "detected a deletion at position 414\n",
      "detected a deletion at position 415\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 417\n",
      "detected a deletion at position 418\n",
      "detected a deletion at position 419\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between W and W\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 423\n",
      "detected a deletion at position 424\n",
      "detected a deletion at position 425\n",
      "detected a deletion at position 426\n",
      "detected a deletion at position 427\n",
      "detected a deletion at position 428\n",
      "detected a deletion at position 429\n",
      "detected a deletion at position 430\n",
      "detected a deletion at position 431\n",
      "detected a deletion at position 432\n",
      "detected a deletion at position 433\n",
      "comparing hydrophobicity between L and L\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between N and Q\n",
      "the difference is hydrophobicity is 1.0\n",
      "detected a deletion at position 436\n",
      "detected a deletion at position 437\n",
      "detected a deletion at position 438\n",
      "detected a deletion at position 439\n",
      "detected a deletion at position 440\n",
      "detected a deletion at position 441\n",
      "detected a deletion at position 442\n",
      "detected a deletion at position 443\n",
      "detected a deletion at position 444\n",
      "detected a deletion at position 445\n",
      "detected a deletion at position 446\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 448\n",
      "detected a deletion at position 449\n",
      "detected a deletion at position 450\n",
      "detected a deletion at position 451\n",
      "detected a deletion at position 452\n",
      "comparing hydrophobicity between Y and Y\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 454\n",
      "detected a deletion at position 455\n",
      "detected a deletion at position 456\n",
      "detected a deletion at position 457\n",
      "detected a deletion at position 458\n",
      "detected a deletion at position 459\n",
      "detected a deletion at position 460\n",
      "comparing hydrophobicity between M and M\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 462\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between A and A\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 466\n",
      "detected a deletion at position 467\n",
      "detected a deletion at position 468\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 470\n",
      "detected a deletion at position 471\n",
      "* position D471 is a residue of interest!\n",
      "comparing hydrophobicity between D and D\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 473\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 475\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between D and D\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between S and S\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between P and P\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between L and L\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 483\n",
      "comparing hydrophobicity between T and C\n",
      "the difference is hydrophobicity is 5.0\n",
      "detected a deletion at position 485\n",
      "detected a deletion at position 486\n",
      "detected a deletion at position 487\n",
      "detected a deletion at position 488\n",
      "detected a deletion at position 489\n",
      "comparing hydrophobicity between W and W\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 491\n",
      "comparing hydrophobicity between L and L\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 493\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 495\n",
      "detected a deletion at position 496\n",
      "comparing hydrophobicity between S and S\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between W and W\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 500\n",
      "detected a deletion at position 501\n",
      "comparing hydrophobicity between C and C\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 503\n",
      "detected a deletion at position 504\n",
      "detected a deletion at position 505\n",
      "detected a deletion at position 506\n",
      "detected a deletion at position 507\n",
      "comparing hydrophobicity between P and P\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and G\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between V and V\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between Y and Y\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between G and T\n",
      "the difference is hydrophobicity is 1.0\n",
      "detected a deletion at position 513\n",
      "comparing hydrophobicity between V and V\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 515\n",
      "detected a deletion at position 516\n",
      "detected a deletion at position 517\n",
      "detected a deletion at position 518\n",
      "detected a deletion at position 519\n",
      "comparing hydrophobicity between W and W\n",
      "the difference is hydrophobicity is 0.0\n",
      "comparing hydrophobicity between I and I\n",
      "the difference is hydrophobicity is 0.0\n",
      "detected a deletion at position 522\n",
      "detected a deletion at position 523\n",
      "detected a deletion at position 524\n",
      "detected a deletion at position 525\n",
      "detected a deletion at position 526\n",
      "detected a deletion at position 527\n",
      "detected a deletion at position 528\n",
      "detected a deletion at position 529\n"
     ]
    }
   ],
   "source": [
    "# we want to keep track of which amino acid our\n",
    "# \"cursor\" is on in the for loop\n",
    "position_counter = 0\n",
    "\n",
    "# get the entire list of positions in the human sequence as\n",
    "# integers. We include dashes in this calculation\n",
    "list_of_positions_including_dashes = range(len(human_compact))\n",
    "\n",
    "for position_with_dashes in list_of_positions_including_dashes:\n",
    "    \n",
    "    # get the amino acid at this position (dashes included)\n",
    "    # in both human and cat\n",
    "    resi_in_human = human_compact[position_with_dashes]\n",
    "    resi_in_cat = cat_record_compact[position_with_dashes]\n",
    "    \n",
    "    # skip this position if it is a '-'\n",
    "    # in the human sequence record\n",
    "    if resi_in_human == '-':\n",
    "        continue\n",
    "    elif position_counter in resi_interest:\n",
    "        # detect if we are at an important amino acid\n",
    "        print(f\"* position {resi_in_human}{position_counter} is a residue of interest!\")\n",
    "        position_counter += 1\n",
    "    else:\n",
    "        # increment the counter by 1\n",
    "        position_counter += 1\n",
    "    \n",
    "    # detect amino acid deletions\n",
    "    if resi_in_cat == '-':\n",
    "        print(f'detected a deletion at position {position_counter}')\n",
    "        continue\n",
    "    \n",
    "    # check changes in amino acid properties\n",
    "    if is_change_in_hydrophobicity(resi_in_human, resi_in_cat):\n",
    "        print(f\"detected a change in hydrophobicity at position {position_counter}\")\n",
    "        \n",
    "    # TODO: check for other changes in amino acid properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Goal for the end of this week\n",
    "\n",
    "For every position in the human sequence (compared to cat sequence), write an algorithm that prints every time there is a hydrophobic residue in human, and non-hydrophobic (hydrophilic) residue in cats."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Other useful `SummaryInfo` tools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute replacement dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('-', 'A'): 35.0,\n",
       " ('-', 'C'): 24.0,\n",
       " ('-', 'D'): 34.0,\n",
       " ('-', 'E'): 39.0,\n",
       " ('-', 'F'): 22.0,\n",
       " ('-', 'G'): 53.0,\n",
       " ('-', 'H'): 23.0,\n",
       " ('-', 'I'): 28.0,\n",
       " ('-', 'K'): 17.0,\n",
       " ('-', 'L'): 50.0,\n",
       " ('-', 'M'): 7.0,\n",
       " ('-', 'N'): 24.0,\n",
       " ('-', 'P'): 75.0,\n",
       " ('-', 'Q'): 25.0,\n",
       " ('-', 'R'): 31.0,\n",
       " ('-', 'S'): 58.0,\n",
       " ('-', 'T'): 29.0,\n",
       " ('-', 'V'): 29.0,\n",
       " ('-', 'W'): 12.0,\n",
       " ('-', 'Y'): 19.0,\n",
       " ('A', '-'): 31.0,\n",
       " ('A', 'M'): 1.0,\n",
       " ('C', '-'): 9.0,\n",
       " ('C', 'L'): 1.0,\n",
       " ('D', '-'): 17.0,\n",
       " ('D', 'S'): 1.0,\n",
       " ('E', '-'): 18.0,\n",
       " ('E', 'G'): 1.0,\n",
       " ('E', 'S'): 1.0,\n",
       " ('F', '-'): 11.0,\n",
       " ('G', '-'): 26.0,\n",
       " ('G', 'H'): 1.0,\n",
       " ('G', 'S'): 1.0,\n",
       " ('G', 'T'): 1.0,\n",
       " ('H', '-'): 7.0,\n",
       " ('H', 'L'): 1.0,\n",
       " ('H', 'V'): 1.0,\n",
       " ('I', '-'): 16.0,\n",
       " ('I', 'A'): 1.0,\n",
       " ('I', 'L'): 2.0,\n",
       " ('I', 'P'): 1.0,\n",
       " ('K', '-'): 22.0,\n",
       " ('K', 'G'): 2.0,\n",
       " ('K', 'L'): 1.0,\n",
       " ('L', '-'): 28.0,\n",
       " ('L', 'P'): 1.0,\n",
       " ('M', '-'): 10.0,\n",
       " ('M', 'L'): 1.0,\n",
       " ('N', '-'): 30.0,\n",
       " ('N', 'L'): 1.0,\n",
       " ('N', 'Q'): 2.0,\n",
       " ('P', '-'): 26.0,\n",
       " ('P', 'C'): 1.0,\n",
       " ('P', 'R'): 1.0,\n",
       " ('P', 'S'): 1.0,\n",
       " ('P', 'T'): 1.0,\n",
       " ('P', 'V'): 1.0,\n",
       " ('Q', '-'): 21.0,\n",
       " ('R', '-'): 15.0,\n",
       " ('R', 'P'): 1.0,\n",
       " ('S', '-'): 42.0,\n",
       " ('S', 'C'): 1.0,\n",
       " ('S', 'G'): 1.0,\n",
       " ('S', 'P'): 1.0,\n",
       " ('T', '-'): 22.0,\n",
       " ('T', 'C'): 1.0,\n",
       " ('T', 'P'): 1.0,\n",
       " ('T', 'W'): 1.0,\n",
       " ('V', '-'): 35.0,\n",
       " ('V', 'C'): 1.0,\n",
       " ('V', 'L'): 1.0,\n",
       " ('W', '-'): 9.0,\n",
       " ('Y', '-'): 22.0}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hum_cat_rep_dict = hum_cat_summary.replacement_dictionary()\n",
    "{k: hum_cat_rep_dict[k] for k in hum_cat_rep_dict\n",
    " if hum_cat_rep_dict[k] > 0\n",
    " and k[0] != k[1]}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute substitution and log odds matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('-', '-'): 3980.0,\n",
       " ('-', 'A'): 66.0,\n",
       " ('-', 'C'): 33.0,\n",
       " ('-', 'D'): 51.0,\n",
       " ('-', 'E'): 57.0,\n",
       " ('-', 'F'): 33.0,\n",
       " ('-', 'G'): 79.0,\n",
       " ('-', 'H'): 30.0,\n",
       " ('-', 'I'): 44.0,\n",
       " ('-', 'K'): 39.0,\n",
       " ('-', 'L'): 78.0,\n",
       " ('-', 'M'): 17.0,\n",
       " ('-', 'N'): 54.0,\n",
       " ('-', 'P'): 101.0,\n",
       " ('-', 'Q'): 46.0,\n",
       " ('-', 'R'): 46.0,\n",
       " ('-', 'S'): 100.0,\n",
       " ('-', 'T'): 51.0,\n",
       " ('-', 'V'): 64.0,\n",
       " ('-', 'W'): 21.0,\n",
       " ('-', 'Y'): 41.0,\n",
       " ('A', 'A'): 5.0,\n",
       " ('A', 'C'): 0,\n",
       " ('A', 'D'): 0,\n",
       " ('A', 'E'): 0,\n",
       " ('A', 'F'): 0,\n",
       " ('A', 'G'): 0,\n",
       " ('A', 'H'): 0,\n",
       " ('A', 'I'): 1.0,\n",
       " ('A', 'K'): 0,\n",
       " ('A', 'L'): 0,\n",
       " ('A', 'M'): 1.0,\n",
       " ('A', 'N'): 0,\n",
       " ('A', 'P'): 0,\n",
       " ('A', 'Q'): 0,\n",
       " ('A', 'R'): 0,\n",
       " ('A', 'S'): 0,\n",
       " ('A', 'T'): 0,\n",
       " ('A', 'V'): 0,\n",
       " ('A', 'W'): 0,\n",
       " ('A', 'Y'): 0,\n",
       " ('C', 'C'): 13.0,\n",
       " ('C', 'D'): 0,\n",
       " ('C', 'E'): 0,\n",
       " ('C', 'F'): 0,\n",
       " ('C', 'G'): 0,\n",
       " ('C', 'H'): 0,\n",
       " ('C', 'I'): 0,\n",
       " ('C', 'K'): 0,\n",
       " ('C', 'L'): 1.0,\n",
       " ('C', 'M'): 0,\n",
       " ('C', 'N'): 0,\n",
       " ('C', 'P'): 1.0,\n",
       " ('C', 'Q'): 0,\n",
       " ('C', 'R'): 0,\n",
       " ('C', 'S'): 1.0,\n",
       " ('C', 'T'): 1.0,\n",
       " ('C', 'V'): 1.0,\n",
       " ('C', 'W'): 0,\n",
       " ('C', 'Y'): 0,\n",
       " ('D', 'D'): 3.0,\n",
       " ('D', 'E'): 0,\n",
       " ('D', 'F'): 0,\n",
       " ('D', 'G'): 0,\n",
       " ('D', 'H'): 0,\n",
       " ('D', 'I'): 0,\n",
       " ('D', 'K'): 0,\n",
       " ('D', 'L'): 0,\n",
       " ('D', 'M'): 0,\n",
       " ('D', 'N'): 0,\n",
       " ('D', 'P'): 0,\n",
       " ('D', 'Q'): 0,\n",
       " ('D', 'R'): 0,\n",
       " ('D', 'S'): 1.0,\n",
       " ('D', 'T'): 0,\n",
       " ('D', 'V'): 0,\n",
       " ('D', 'W'): 0,\n",
       " ('D', 'Y'): 0,\n",
       " ('E', 'E'): 0,\n",
       " ('E', 'F'): 0,\n",
       " ('E', 'G'): 1.0,\n",
       " ('E', 'H'): 0,\n",
       " ('E', 'I'): 0,\n",
       " ('E', 'K'): 0,\n",
       " ('E', 'L'): 0,\n",
       " ('E', 'M'): 0,\n",
       " ('E', 'N'): 0,\n",
       " ('E', 'P'): 0,\n",
       " ('E', 'Q'): 0,\n",
       " ('E', 'R'): 0,\n",
       " ('E', 'S'): 1.0,\n",
       " ('E', 'T'): 0,\n",
       " ('E', 'V'): 0,\n",
       " ('E', 'W'): 0,\n",
       " ('E', 'Y'): 0,\n",
       " ('F', 'F'): 0,\n",
       " ('F', 'G'): 0,\n",
       " ('F', 'H'): 0,\n",
       " ('F', 'I'): 0,\n",
       " ('F', 'K'): 0,\n",
       " ('F', 'L'): 0,\n",
       " ('F', 'M'): 0,\n",
       " ('F', 'N'): 0,\n",
       " ('F', 'P'): 0,\n",
       " ('F', 'Q'): 0,\n",
       " ('F', 'R'): 0,\n",
       " ('F', 'S'): 0,\n",
       " ('F', 'T'): 0,\n",
       " ('F', 'V'): 0,\n",
       " ('F', 'W'): 0,\n",
       " ('F', 'Y'): 0,\n",
       " ('G', 'G'): 19.0,\n",
       " ('G', 'H'): 1.0,\n",
       " ('G', 'I'): 0,\n",
       " ('G', 'K'): 2.0,\n",
       " ('G', 'L'): 0,\n",
       " ('G', 'M'): 0,\n",
       " ('G', 'N'): 0,\n",
       " ('G', 'P'): 0,\n",
       " ('G', 'Q'): 0,\n",
       " ('G', 'R'): 0,\n",
       " ('G', 'S'): 2.0,\n",
       " ('G', 'T'): 1.0,\n",
       " ('G', 'V'): 0,\n",
       " ('G', 'W'): 0,\n",
       " ('G', 'Y'): 0,\n",
       " ('H', 'H'): 3.0,\n",
       " ('H', 'I'): 0,\n",
       " ('H', 'K'): 0,\n",
       " ('H', 'L'): 1.0,\n",
       " ('H', 'M'): 0,\n",
       " ('H', 'N'): 0,\n",
       " ('H', 'P'): 0,\n",
       " ('H', 'Q'): 0,\n",
       " ('H', 'R'): 0,\n",
       " ('H', 'S'): 0,\n",
       " ('H', 'T'): 0,\n",
       " ('H', 'V'): 1.0,\n",
       " ('H', 'W'): 0,\n",
       " ('H', 'Y'): 0,\n",
       " ('I', 'I'): 3.0,\n",
       " ('I', 'K'): 0,\n",
       " ('I', 'L'): 2.0,\n",
       " ('I', 'M'): 0,\n",
       " ('I', 'N'): 0,\n",
       " ('I', 'P'): 1.0,\n",
       " ('I', 'Q'): 0,\n",
       " ('I', 'R'): 0,\n",
       " ('I', 'S'): 0,\n",
       " ('I', 'T'): 0,\n",
       " ('I', 'V'): 0,\n",
       " ('I', 'W'): 0,\n",
       " ('I', 'Y'): 0,\n",
       " ('K', 'K'): 0,\n",
       " ('K', 'L'): 1.0,\n",
       " ('K', 'M'): 0,\n",
       " ('K', 'N'): 0,\n",
       " ('K', 'P'): 0,\n",
       " ('K', 'Q'): 0,\n",
       " ('K', 'R'): 0,\n",
       " ('K', 'S'): 0,\n",
       " ('K', 'T'): 0,\n",
       " ('K', 'V'): 0,\n",
       " ('K', 'W'): 0,\n",
       " ('K', 'Y'): 0,\n",
       " ('L', 'L'): 7.0,\n",
       " ('L', 'M'): 1.0,\n",
       " ('L', 'N'): 1.0,\n",
       " ('L', 'P'): 1.0,\n",
       " ('L', 'Q'): 0,\n",
       " ('L', 'R'): 0,\n",
       " ('L', 'S'): 0,\n",
       " ('L', 'T'): 0,\n",
       " ('L', 'V'): 1.0,\n",
       " ('L', 'W'): 0,\n",
       " ('L', 'Y'): 0,\n",
       " ('M', 'M'): 1.0,\n",
       " ('M', 'N'): 0,\n",
       " ('M', 'P'): 0,\n",
       " ('M', 'Q'): 0,\n",
       " ('M', 'R'): 0,\n",
       " ('M', 'S'): 0,\n",
       " ('M', 'T'): 0,\n",
       " ('M', 'V'): 0,\n",
       " ('M', 'W'): 0,\n",
       " ('M', 'Y'): 0,\n",
       " ('N', 'N'): 0,\n",
       " ('N', 'P'): 0,\n",
       " ('N', 'Q'): 2.0,\n",
       " ('N', 'R'): 0,\n",
       " ('N', 'S'): 0,\n",
       " ('N', 'T'): 0,\n",
       " ('N', 'V'): 0,\n",
       " ('N', 'W'): 0,\n",
       " ('N', 'Y'): 0,\n",
       " ('P', 'P'): 7.0,\n",
       " ('P', 'Q'): 0,\n",
       " ('P', 'R'): 2.0,\n",
       " ('P', 'S'): 2.0,\n",
       " ('P', 'T'): 2.0,\n",
       " ('P', 'V'): 1.0,\n",
       " ('P', 'W'): 0,\n",
       " ('P', 'Y'): 0,\n",
       " ('Q', 'Q'): 1.0,\n",
       " ('Q', 'R'): 0,\n",
       " ('Q', 'S'): 0,\n",
       " ('Q', 'T'): 0,\n",
       " ('Q', 'V'): 0,\n",
       " ('Q', 'W'): 0,\n",
       " ('Q', 'Y'): 0,\n",
       " ('R', 'R'): 1.0,\n",
       " ('R', 'S'): 0,\n",
       " ('R', 'T'): 0,\n",
       " ('R', 'V'): 0,\n",
       " ('R', 'W'): 0,\n",
       " ('R', 'Y'): 0,\n",
       " ('S', 'S'): 3.0,\n",
       " ('S', 'T'): 0,\n",
       " ('S', 'V'): 0,\n",
       " ('S', 'W'): 0,\n",
       " ('S', 'Y'): 0,\n",
       " ('T', 'T'): 0,\n",
       " ('T', 'V'): 0,\n",
       " ('T', 'W'): 1.0,\n",
       " ('T', 'Y'): 0,\n",
       " ('V', 'V'): 2.0,\n",
       " ('V', 'W'): 0,\n",
       " ('V', 'Y'): 0,\n",
       " ('W', 'W'): 6.0,\n",
       " ('W', 'Y'): 0,\n",
       " ('Y', 'Y'): 2.0}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_arm = SubsMat.SeqMat(hum_cat_rep_dict)\n",
    "my_arm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('-', '-'): 0.012004147,\n",
       " ('-', 'A'): -0.050085165,\n",
       " ('A', 'A'): 4.079518096,\n",
       " ('-', 'C'): -0.764682946,\n",
       " ('A', 'C'): -999,\n",
       " ('C', 'C'): 6.028834156,\n",
       " ('-', 'D'): 0.005367282,\n",
       " ('A', 'D'): -999,\n",
       " ('C', 'D'): -999,\n",
       " ('D', 'D'): 4.197394949,\n",
       " ('-', 'E'): 0.1411699,\n",
       " ('A', 'E'): -999,\n",
       " ('C', 'E'): -999,\n",
       " ('D', 'E'): -999,\n",
       " ('E', 'E'): -999,\n",
       " ('-', 'F'): 0.190922935,\n",
       " ('A', 'F'): -999,\n",
       " ('C', 'F'): -999,\n",
       " ('D', 'F'): -999,\n",
       " ('E', 'F'): -999,\n",
       " ('F', 'F'): -999,\n",
       " ('-', 'G'): -0.459492627,\n",
       " ('A', 'G'): -999,\n",
       " ('C', 'G'): -999,\n",
       " ('D', 'G'): -999,\n",
       " ('E', 'G'): 0.491555079,\n",
       " ('F', 'G'): -999,\n",
       " ('G', 'G'): 4.667929331,\n",
       " ('-', 'H'): -0.187588688,\n",
       " ('A', 'H'): -999,\n",
       " ('C', 'H'): -999,\n",
       " ('D', 'H'): -999,\n",
       " ('E', 'H'): -999,\n",
       " ('F', 'H'): -999,\n",
       " ('G', 'H'): 1.088795909,\n",
       " ('H', 'H'): 5.342552501,\n",
       " ('-', 'I'): -0.104532949,\n",
       " ('A', 'I'): 1.288104717,\n",
       " ('C', 'I'): -999,\n",
       " ('D', 'I'): -999,\n",
       " ('E', 'I'): -999,\n",
       " ('F', 'I'): -999,\n",
       " ('G', 'I'): -999,\n",
       " ('H', 'I'): -999,\n",
       " ('I', 'I'): 4.403581935,\n",
       " ('-', 'K'): 0.084007731,\n",
       " ('A', 'K'): -999,\n",
       " ('C', 'K'): -999,\n",
       " ('D', 'K'): -999,\n",
       " ('E', 'K'): -999,\n",
       " ('F', 'K'): -999,\n",
       " ('G', 'K'): 1.981880705,\n",
       " ('H', 'K'): -999,\n",
       " ('I', 'K'): -999,\n",
       " ('K', 'K'): -999,\n",
       " ('-', 'L'): -0.181886329,\n",
       " ('A', 'L'): -999,\n",
       " ('C', 'L'): 0.670182956,\n",
       " ('D', 'L'): -999,\n",
       " ('E', 'L'): -999,\n",
       " ('F', 'L'): -999,\n",
       " ('G', 'L'): -999,\n",
       " ('H', 'L'): 1.384780737,\n",
       " ('I', 'L'): 1.915295453,\n",
       " ('K', 'L'): 1.277865533,\n",
       " ('L', 'L'): 3.819326395,\n",
       " ('-', 'M'): -0.113931647,\n",
       " ('A', 'M'): 2.650674797,\n",
       " ('C', 'M'): -999,\n",
       " ('D', 'M'): -999,\n",
       " ('E', 'M'): -999,\n",
       " ('F', 'M'): -999,\n",
       " ('G', 'M'): -999,\n",
       " ('H', 'M'): -999,\n",
       " ('I', 'M'): -999,\n",
       " ('K', 'M'): -999,\n",
       " ('L', 'M'): 2.277865533,\n",
       " ('M', 'M'): 5.543759593,\n",
       " ('-', 'N'): 0.112920423,\n",
       " ('A', 'N'): -999,\n",
       " ('C', 'N'): -999,\n",
       " ('D', 'N'): -999,\n",
       " ('E', 'N'): -999,\n",
       " ('F', 'N'): -999,\n",
       " ('G', 'N'): -999,\n",
       " ('H', 'N'): -999,\n",
       " ('I', 'N'): -999,\n",
       " ('K', 'N'): -999,\n",
       " ('L', 'N'): 0.837292941,\n",
       " ('M', 'N'): -999,\n",
       " ('N', 'N'): -999,\n",
       " ('-', 'P'): -0.116649867,\n",
       " ('A', 'P'): -999,\n",
       " ('C', 'P'): 0.362610154,\n",
       " ('D', 'P'): -999,\n",
       " ('E', 'P'): -999,\n",
       " ('F', 'P'): -999,\n",
       " ('G', 'P'): -999,\n",
       " ('H', 'P'): -999,\n",
       " ('I', 'P'): 0.607722652,\n",
       " ('K', 'P'): -999,\n",
       " ('L', 'P'): -0.295601329,\n",
       " ('M', 'P'): -999,\n",
       " ('N', 'P'): -999,\n",
       " ('P', 'P'): 3.204180791,\n",
       " ('-', 'Q'): 0.070628701,\n",
       " ('A', 'Q'): -999,\n",
       " ('C', 'Q'): -999,\n",
       " ('D', 'Q'): -999,\n",
       " ('E', 'Q'): -999,\n",
       " ('F', 'Q'): -999,\n",
       " ('G', 'Q'): -999,\n",
       " ('H', 'Q'): -999,\n",
       " ('I', 'Q'): -999,\n",
       " ('K', 'Q'): -999,\n",
       " ('L', 'Q'): -999,\n",
       " ('M', 'Q'): -999,\n",
       " ('N', 'Q'): 2.851648234,\n",
       " ('P', 'Q'): -999,\n",
       " ('Q', 'Q'): 3.040682059,\n",
       " ('-', 'R'): 0.070628701,\n",
       " ('A', 'R'): -999,\n",
       " ('C', 'R'): -999,\n",
       " ('D', 'R'): -999,\n",
       " ('E', 'R'): -999,\n",
       " ('F', 'R'): -999,\n",
       " ('G', 'R'): -999,\n",
       " ('H', 'R'): -999,\n",
       " ('I', 'R'): -999,\n",
       " ('K', 'R'): -999,\n",
       " ('L', 'R'): -999,\n",
       " ('M', 'R'): -999,\n",
       " ('N', 'R'): -999,\n",
       " ('P', 'R'): 1.718753964,\n",
       " ('Q', 'R'): -999,\n",
       " ('R', 'R'): 3.040682059,\n",
       " ('-', 'S'): 0.014600162,\n",
       " ('A', 'S'): -999,\n",
       " ('C', 'S'): 0.508215476,\n",
       " ('D', 'S'): 0.650234481,\n",
       " ('E', 'S'): 0.625572427,\n",
       " ('F', 'S'): -999,\n",
       " ('G', 'S'): 0.554019166,\n",
       " ('H', 'S'): -999,\n",
       " ('I', 'S'): -999,\n",
       " ('K', 'S'): -999,\n",
       " ('L', 'S'): -999,\n",
       " ('M', 'S'): -999,\n",
       " ('N', 'S'): -999,\n",
       " ('P', 'S'): 0.542431191,\n",
       " ('Q', 'S'): -999,\n",
       " ('R', 'S'): -999,\n",
       " ('S', 'S'): 2.272999014,\n",
       " ('-', 'T'): 0.055993355,\n",
       " ('A', 'T'): -999,\n",
       " ('C', 'T'): 1.521039516,\n",
       " ('D', 'T'): -999,\n",
       " ('E', 'T'): -999,\n",
       " ('F', 'T'): -999,\n",
       " ('G', 'T'): 0.566843206,\n",
       " ('H', 'T'): -999,\n",
       " ('I', 'T'): -999,\n",
       " ('K', 'T'): -999,\n",
       " ('L', 'T'): -999,\n",
       " ('M', 'T'): -999,\n",
       " ('N', 'T'): -999,\n",
       " ('P', 'T'): 1.555255232,\n",
       " ('Q', 'T'): -999,\n",
       " ('R', 'T'): -999,\n",
       " ('S', 'T'): -999,\n",
       " ('T', 'T'): -999,\n",
       " ('-', 'V'): 0.020997934,\n",
       " ('A', 'V'): -999,\n",
       " ('C', 'V'): 1.158469437,\n",
       " ('D', 'V'): -999,\n",
       " ('E', 'V'): -999,\n",
       " ('F', 'V'): -999,\n",
       " ('G', 'V'): -999,\n",
       " ('H', 'V'): 1.873067218,\n",
       " ('I', 'V'): -999,\n",
       " ('K', 'V'): -999,\n",
       " ('L', 'V'): 0.500257954,\n",
       " ('M', 'V'): -999,\n",
       " ('N', 'V'): -999,\n",
       " ('P', 'V'): 0.192685152,\n",
       " ('Q', 'V'): -999,\n",
       " ('R', 'V'): -999,\n",
       " ('S', 'V'): -999,\n",
       " ('T', 'V'): -999,\n",
       " ('V', 'V'): 2.988544435,\n",
       " ('-', 'W'): -0.504222483,\n",
       " ('A', 'W'): -999,\n",
       " ('C', 'W'): -999,\n",
       " ('D', 'W'): -999,\n",
       " ('E', 'W'): -999,\n",
       " ('F', 'W'): -999,\n",
       " ('G', 'W'): -999,\n",
       " ('H', 'W'): -999,\n",
       " ('I', 'W'): -999,\n",
       " ('K', 'W'): -999,\n",
       " ('L', 'W'): -999,\n",
       " ('M', 'W'): -999,\n",
       " ('N', 'W'): -999,\n",
       " ('P', 'W'): -999,\n",
       " ('Q', 'W'): -999,\n",
       " ('R', 'W'): -999,\n",
       " ('S', 'W'): -999,\n",
       " ('T', 'W'): 2.433576675,\n",
       " ('V', 'W'): -999,\n",
       " ('W', 'W'): 6.738431257,\n",
       " ('-', 'Y'): 0.056621843,\n",
       " ('A', 'Y'): -999,\n",
       " ('C', 'Y'): -999,\n",
       " ('D', 'Y'): -999,\n",
       " ('E', 'Y'): -999,\n",
       " ('F', 'Y'): -999,\n",
       " ('G', 'Y'): -999,\n",
       " ('H', 'Y'): -999,\n",
       " ('I', 'Y'): -999,\n",
       " ('K', 'Y'): -999,\n",
       " ('L', 'Y'): -999,\n",
       " ('M', 'Y'): -999,\n",
       " ('N', 'Y'): -999,\n",
       " ('P', 'Y'): -999,\n",
       " ('Q', 'Y'): -999,\n",
       " ('R', 'Y'): -999,\n",
       " ('S', 'Y'): -999,\n",
       " ('T', 'Y'): -999,\n",
       " ('V', 'Y'): -999,\n",
       " ('W', 'Y'): -999,\n",
       " ('Y', 'Y'): 4.344688246}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_lom = SubsMat.make_log_odds_matrix(my_arm)\n",
    "my_lom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ty's Work Below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unexpected indent (<ipython-input-53-492dc3e39b13>, line 10)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-53-492dc3e39b13>\"\u001b[0;36m, line \u001b[0;32m10\u001b[0m\n\u001b[0;31m    for position[1][resi_in_human] > 1 in hum_mouse_pssm.pssm:\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unexpected indent\n"
     ]
    }
   ],
   "source": [
    "for position in hum_mouse_pssm.pssm:\n",
    "    resi_in_human = position[0]\n",
    "    resi_dict=position[1]\n",
    "    if resi_in_human ==\"-\": \n",
    "        continue\n",
    "    else:\n",
    "        position_counter += 1\n",
    "    if position[1][resi_in_human] > 1:\n",
    "        print(f\"mouse and human are the same at position\" + f\"{position_counter}, which is amino acid {resi_in_human}\")\n",
    "            for position[1][resi_in_human] > 1 in hum_mouse_pssm.pssm:\n",
    "                if position[0] == [\"D\", \"E\"]:\n",
    "                    print(f\"mouse and human have the same acidic amino acid at position\" + f\"{position_counter}, which is amino {resi_in_human}\")\n",
    "                elif position[0]==[\"R\", \"H\", \"K\"]:\n",
    "                    Print(f\"mouse and human have the same basic amino acid at position\" + f\"{position_counter}, which is amino acid {resi_in_human}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "position_counter=0\n",
    "for position in hum_mouse_pssm.pssm:\n",
    "    resi_in_human= position[0]\n",
    "    resi_dict= position[1]\n",
    "    if resi_in_human == '-':\n",
    "        continue\n",
    "    else:\n",
    "        position_counter += 1\n",
    "    if position[1][resi_in_human] > 1:\n",
    "        print(f\"mouse and human are the same at position\" + f\"{position_counter}, which is amino acid {resi_in_human}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to do all the species automatically"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Define a function `compute_sequence_diff` that (basically) has all the code in this notebook\n",
    "2. Use a function\n",
    "\n",
    "```python\n",
    "for species in species_list:\n",
    "    compute_sequence_diff('Homo sapiens', 'monkey')\n",
    "    \n",
    "def compute_sequence_diff(species1, species2):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # do stuff with species sequence\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
