{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 311,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numbers import Number\n",
    "from Bio.Align import AlignInfo, MultipleSeqAlignment\n",
    "from Bio import AlignIO, Alphabet, SeqRecord, Seq, SubsMat\n",
    "from datetime import datetime as dt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Similarity Analysis\n",
    "----\n",
    "This notebook is a follow-up of the [sequence similarity demo](./20200706_seq_sim.ipynb). In this analysis, we will answer the question:\n",
    "\n",
    "_How does the primary sequence of TMPRSS2 differ between species that one would encounter in a farm environment?_\n",
    "\n",
    "It specifically provides a solution for the below analysis:\n",
    "\n",
    "* Repeat analysis for each of the other domestic species (dog, horse, chicken, etc.)\n",
    "* Generate a \"generalized PSSM\" for the other types of penalized polymorphisms, such as `acidic -> basic`, `bulky -> small`, `aromatic -> non-aromatic`, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pseudocode summary\n",
    "----\n",
    "\n",
    "## Analysis Pipeline Overview\n",
    "\n",
    "1. Read in all aligned sequences in FASTA file\n",
    "2. Filter aligned sequences to only species of interest\n",
    "3. **Optimal**: somehow remove proteins that are obviously not TMPRSS2\n",
    "    - This would be some sort of function\n",
    "4. Generate a mapping of comparisons, e.g. `[['human', 'mouse'], ['human', 'cattle'], ...]`\n",
    "5. Get consensus sequence for species with more than 1 sequence, using function `get_consensus`\n",
    "5. For each pair of sequences in the above mapping, call `compare_resi_props`\n",
    "    - This function will contain much of the analysis in [sequence similarity demo](./20200706_seq_sim.ipynb), generalized for any two sequences. \n",
    "    - Returns some sort of data structure (probably a dataframe, we will call it `resi_prop_delta`) that shows change in each amino acid property (hydrophobicity, charge, etc.) for each site in the protein sequence.\n",
    "6. Pass each `resi_prop_delta` to `get_struc_similarity`\n",
    "    - This function will arbitrarily penalize changes in amino acids on a unitless scale\n",
    "    - Returns a unitless structural similarity score (type float), where lower values indicate that a pair of homologs are functionally similar, and higher values indicate divergence in structure/function.\n",
    "    \n",
    "## `compare_resi_props` function\n",
    "- Arguments\n",
    "    - Two `Seq` instances\n",
    "- Returns\n",
    "    - Pandas dataframe `resi_prop_delta`: integer index at each site, each column is change in residue property for that site\n",
    "    \n",
    "## `get_struc_similarity` function\n",
    "- Arguments\n",
    "    - File path to `resi_prop_delta` CSV\n",
    "- Returns\n",
    "    - Float"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Read in all aligned sequences in FASTA file\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<<class 'Bio.Align.MultipleSeqAlignment'> instance (9757 records of length 60918, SingleLetterAlphabet()) at 7fed8a9f5590>"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment = AlignIO.read(open('./trimmed_alg.txt'), format='fasta')\n",
    "alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Filter aligned sequences to only species of interest\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "domestic_sp_names = [\n",
    "    'Homo sapiens', # human\n",
    "    'Mus musculus', # mouse\n",
    "    'Canis lupus familiaris', # dog\n",
    "    'Felis catus', # cat\n",
    "    'Bos taurus', # cattle\n",
    "    'Equus caballus', # horse\n",
    "    'Gallus gallus' # chicken\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_1</th>\n",
       "      <th>id_2</th>\n",
       "      <th>species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CRE24749</td>\n",
       "      <td>CRE24749</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:DS268562,E3N945_CAERE,E3N945,CRE_24749...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRE21132</td>\n",
       "      <td>CRE-TRY-4</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:E3MEX0,DS268440,E3MEX0_CAERE,CRE21132,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRE24758</td>\n",
       "      <td>CRE-TRY-6</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:E3N963,DS268562,E3N963_CAERE,CRE24758,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRE18672</td>\n",
       "      <td>CRE18672</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:DS268410,E3LKX4_CAERE,E3LKX4,CRE_18672...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRE24729</td>\n",
       "      <td>CRE-TRY-3</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>31234</td>\n",
       "      <td>aliases:E3N418,DS268522,E3N418_CAERE,CRE24729,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id_1       id_2                 species         \\\n",
       "0  CRE24749   CRE24749  Caenorhabditis remanei  31234   \n",
       "1  CRE21132  CRE-TRY-4  Caenorhabditis remanei  31234   \n",
       "2  CRE24758  CRE-TRY-6  Caenorhabditis remanei  31234   \n",
       "3  CRE18672   CRE18672  Caenorhabditis remanei  31234   \n",
       "4  CRE24729  CRE-TRY-3  Caenorhabditis remanei  31234   \n",
       "\n",
       "                                                      \n",
       "0  aliases:DS268562,E3N945_CAERE,E3N945,CRE_24749...  \n",
       "1  aliases:E3MEX0,DS268440,E3MEX0_CAERE,CRE21132,...  \n",
       "2  aliases:E3N963,DS268562,E3N963_CAERE,CRE24758,...  \n",
       "3  aliases:DS268410,E3LKX4_CAERE,E3LKX4,CRE_18672...  \n",
       "4  aliases:E3N418,DS268522,E3N418_CAERE,CRE24729,...  "
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmprss2_ext = pd.read_table('../seq_sim_demo/extended_members.txt', header=None)\n",
    "tmprss2_ext.columns = ['id_1', 'id_2', 'species', '', '']\n",
    "tmprss2_ext.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in alignment:\n",
    "    \n",
    "    # while we're at it, let's make sure that Biopython knows these\n",
    "    # are protein sequences\n",
    "    record.seq.alphabet = Alphabet.generic_protein\n",
    "    \n",
    "    # from visual inspection we know the name format is XXXX.unique_id,\n",
    "    # so we split on \".\" and take the last element of the list\n",
    "    id_code = record.id.split('.')[-1]\n",
    "    \n",
    "    # reference the metadata to get the species name\n",
    "    sp_name = tmprss2_ext[tmprss2_ext['id_1'] == id_code]['species'].values\n",
    "    \n",
    "    try:\n",
    "        sp_name = sp_name.item()\n",
    "    except ValueError:\n",
    "        sp_name = None\n",
    "    \n",
    "    # assign the species name to the species attribute\n",
    "    record.description = sp_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [],
   "source": [
    "dom_aln_list = [record for record in alignment\n",
    "                if record.description in domestic_sp_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<<class 'Bio.Align.MultipleSeqAlignment'> instance (732 records of length 60918, ProteinAlphabet()) at 7fed899918d0>"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dom_aln = MultipleSeqAlignment(dom_aln_list)\n",
    "dom_aln"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Filter proteins that are not TMPRSS2\n",
    "----\n",
    "WIP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Generate mapping of comparisons to make\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "comp_map = [\n",
    "    ['Homo sapiens', 'Mus musculus'],\n",
    "    ['Homo sapiens', 'Canis lupus familiaris'],\n",
    "    ['Homo sapiens', 'Felis catus'],\n",
    "    ['Homo sapiens', 'Bos taurus'],\n",
    "    ['Homo sapiens', 'Equus caballus'],\n",
    "    ['Homo sapiens', 'Gallus gallus' ]\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Get the sequence of human isoform 2\n",
    "----\n",
    "\n",
    "Let's find the sequence record that has the same sequence as isoform 2 on the [TMPRSS2 UniProt page](https://www.uniprot.org/uniprot/O15393#O15393-1). The first few residues of this isoform are `MPPAPPGG`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_aln_list = [\n",
    "    record for record in dom_aln\n",
    "    if record.description == 'Homo sapiens'\n",
    "]\n",
    "human_aln = MultipleSeqAlignment(human_aln_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "isoform_aln_list = [\n",
    "    record for record in human_aln\n",
    "    if 'MPPAPPGG' in str(record.seq).replace(\"-\", \"\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of human sequences that contain MPPAPPGG: 1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('------------------------------------------------------...---', ProteinAlphabet()), id='9606.ENSP00000381588', name='9606.ENSP00000381588', description='Homo sapiens', dbxrefs=[])"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"number of human sequences that contain MPPAPPGG:\", len(isoform_aln_list))\n",
    "human_iso2 = isoform_aln_list[0]\n",
    "human_iso2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also notice that most of the sequence of interest is in the middle of the aligned sequence. Let's trim the aligned sequence to generate a compact aligned sequence that it starts with `MPPAPP` and ends with `ADG`. To do this, we will make use of the [`str.index`](https://docs.python.org/2/library/stdtypes.html?highlight=index#str.index) method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index of N-terminus: 33713\n",
      "index of C-terminus: 38856\n"
     ]
    }
   ],
   "source": [
    "index_nterm = str(human_iso2.seq).index('MPPAPP')\n",
    "index_cterm = str(human_iso2.seq).index('ADG')\n",
    "\n",
    "# since we want to cut at ADG^, not ^ADG, we add 3 characters to this index\n",
    "index_cterm += 3\n",
    "\n",
    "print(\"index of N-terminus:\", index_nterm)\n",
    "print(\"index of C-terminus:\", index_cterm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use these indices to trim to the compact sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "human_compact = human_iso2[index_nterm:index_cterm]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. An alternative to using consensus sequence\n",
    "---\n",
    "Consensus sequences are mostly dashes: how could we fix this?\n",
    "\n",
    "Consensus sequence does a very bad job of choosing sequences that have much other than dashes. Instead, try calculating percent identity, and using the sequence out of a set that has best percent identity to the human reference sequence. The hypothesis here is that this will choose which of the sequences is \"real\" TMPRSS2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_equal(items):\n",
    "    \"\"\"Returns True iff all items are equal.\"\"\"\n",
    "    first = items[0]\n",
    "    return all(x == first for x in items)\n",
    "\n",
    "def all_null(items):\n",
    "    return all(x == '-' for x in items)\n",
    "\n",
    "def percent_identity(aligned_sequences):\n",
    "    \"\"\"Returns the ratio of same-character columns in ``aligned_sequences``.\n",
    "\n",
    "    :param aligned_sequences: a list of strings or equal length.\n",
    "    \"\"\"\n",
    "    match_count = 0\n",
    "    mismatch_count = 0\n",
    "    for chars in zip(*aligned_sequences):\n",
    "        # Here chars is a column of chars, \n",
    "        # one taken from each element of aligned_sequences.\n",
    "        if all_null(chars):\n",
    "            continue\n",
    "        elif all_equal(chars):\n",
    "            match_count += 1\n",
    "        else:\n",
    "            mismatch_count += 1\n",
    "    # return float(match_count) / float(mismatch_count)\n",
    "    # What would make more sense:\n",
    "    return float(match_count) / len(aligned_sequences[0].replace('-', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def get_best_match(msa, untrimmed_ref, trim):\n",
    "    best_perc_id = 0.\n",
    "    best_match = None\n",
    "    for seq in msa:\n",
    "        aligned_sequences = [\n",
    "            str(human_iso2.seq)[trim],\n",
    "            str(seq.seq)[trim]\n",
    "        ]\n",
    "        perc_id = percent_identity(aligned_sequences)\n",
    "        if perc_id > best_perc_id:\n",
    "            best_match = seq\n",
    "            best_perc_id = perc_id\n",
    "    print(f\"assigning sequence with {100 * best_perc_id} percent identity as best\")\n",
    "    return best_match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assigning sequence with 27.22117202268431 percent identity as best\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('------------------------------------------------------...---', ProteinAlphabet()), id='9685.ENSFCAP00000010917', name='9685.ENSFCAP00000010917', description='Felis catus', dbxrefs=[])"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_cat = get_best_match(cat_aln, human_iso2, slice(index_nterm, index_cterm))\n",
    "best_cat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_best_match_by_description(msa, untrimmed_ref, description, trim=slice(0, None)):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    assert isinstance(msa, MultipleSeqAlignment)\n",
    "    \n",
    "    filtered_msa_as_list = [\n",
    "        record for record in msa\n",
    "        if record.description == description\n",
    "    ]\n",
    "    filtered_msa = MultipleSeqAlignment(filtered_msa_as_list)\n",
    "    return get_best_match(filtered_msa, untrimmed_ref, trim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assigning sequence with 2.799922224382656 percent identity as best\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('------------------------------------------------------...---', ProteinAlphabet()), id='9685.ENSFCAP00000010917', name='9685.ENSFCAP00000010917', description='Felis catus', dbxrefs=[])"
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_best_match = get_best_match_by_description(\n",
    "    cat_aln, human_iso2, 'Felis catus',\n",
    "    trim=slice(index_nterm, index_cterm))\n",
    "cat_best_match"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Generate consensus sequences for cat homolog\n",
    "----\n",
    "Or rather, define functions that do this for us :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_consensus(msa, trim=slice(0, None)):\n",
    "    \"\"\"Given a MultipleSequenceAlignment object,\n",
    "    return a consensus sequence. Trim the consensus sequence\n",
    "    using slice `trim`.\n",
    "    \"\"\"\n",
    "    assert isinstance(msa, MultipleSeqAlignment)\n",
    "    \n",
    "    # generate SummaryInfo instance\n",
    "    summary_info = AlignInfo.SummaryInfo(msa)\n",
    "    \n",
    "    # calculate consensus sequence\n",
    "    consensus = summary_info.dumb_consensus(\n",
    "        threshold=0.5, ambiguous='-',\n",
    "        consensus_alpha=None, require_multiple=0)\n",
    "    \n",
    "    # replace X with - and trim of N and C terms\n",
    "    replaced =  str(consensus).replace('X', '-')[trim]\n",
    "    \n",
    "    # some reporting: print number of dashes out of entire length of sequence\n",
    "    num_dashes = replaced.count('-')\n",
    "    print(f\"consensus sequence has {num_dashes} null out of {len(replaced)}\")\n",
    "\n",
    "    # convert to a SeqRecord.SeqRecord instance\n",
    "    compact = SeqRecord.SeqRecord(Seq.Seq(replaced))\n",
    "    return compact"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Example of how to use this function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat_aln_list = [\n",
    "    record for record in dom_aln\n",
    "    if record.description == 'Felis catus'\n",
    "]\n",
    "cat_aln = MultipleSeqAlignment(cat_aln_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "consensus sequence has 4060 null out of 5143\n"
     ]
    }
   ],
   "source": [
    "cat_record_compact = get_consensus(cat_aln, slice(index_nterm, index_cterm))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's generalize further; given a species name, return the consensus sequence as a SequenceRecord:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_consensus_by_description(msa, description, trim=slice(0, None)):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    assert isinstance(msa, MultipleSeqAlignment)\n",
    "    \n",
    "    filtered_msa_as_list = [\n",
    "        record for record in msa\n",
    "        if record.description == description\n",
    "    ]\n",
    "    filtered_msa = MultipleSeqAlignment(filtered_msa_as_list)\n",
    "    return get_consensus(filtered_msa, trim)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using this function is even easier than using `get_consensus`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "consensus sequence has 4060 null out of 5143\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('LSRMP-GL-----C-A-LA----G---A--L---SVALTL--A--P--------...---'), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])"
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_record = get_consensus_by_description(dom_aln, \"Felis catus\", slice(index_nterm, index_cterm))\n",
    "cat_record"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Define functions that compare residues properties\n",
    "---\n",
    "\n",
    "6. For each pair of sequences in the above mapping, call `compare_resi_prop`\n",
    "    - This function will contain much of the analysis in [sequence similarity demo](./20200706_seq_sim.ipynb), generalized for any two sequences. \n",
    "    - Returns some sort of data structure (probably a dataframe, we will call it `resi_prop_delta`) that shows change in each amino acid property (hydrophobicity, charge, etc.) for each site in the protein sequence.\n",
    "7. Pass each `resi_prop_delta` to `get_struc_similarity`\n",
    "    - This function will arbitrarily penalize changes in amino acids on a unitless scale\n",
    "    - Returns a unitless structural similarity score (type float), where lower values indicate that a pair of homologs are functionally similar, and higher values indicate divergence in structure/function."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get table of amino acid properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_name</th>\n",
       "      <th>pka</th>\n",
       "      <th>mass</th>\n",
       "      <th>occurrence</th>\n",
       "      <th>percent_buried</th>\n",
       "      <th>buried_vol</th>\n",
       "      <th>vdw</th>\n",
       "      <th>accessible</th>\n",
       "      <th>hydrophobicity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>single_letter</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>alanine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.08</td>\n",
       "      <td>7.5</td>\n",
       "      <td>38.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>argenine</td>\n",
       "      <td>12.5</td>\n",
       "      <td>156.10</td>\n",
       "      <td>5.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>225.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>asparagine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>114.04</td>\n",
       "      <td>4.6</td>\n",
       "      <td>10.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>aspartic acid</td>\n",
       "      <td>3.9</td>\n",
       "      <td>115.08</td>\n",
       "      <td>5.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>125.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>cysteine</td>\n",
       "      <td>8.3</td>\n",
       "      <td>103.01</td>\n",
       "      <td>1.8</td>\n",
       "      <td>47.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>glutamine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128.06</td>\n",
       "      <td>4.1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>161.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>glutamic acid</td>\n",
       "      <td>4.3</td>\n",
       "      <td>129.12</td>\n",
       "      <td>6.3</td>\n",
       "      <td>20.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>glycine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.05</td>\n",
       "      <td>7.1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>histidine</td>\n",
       "      <td>6.0</td>\n",
       "      <td>137.16</td>\n",
       "      <td>2.2</td>\n",
       "      <td>19.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>isoleucine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.16</td>\n",
       "      <td>5.5</td>\n",
       "      <td>65.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>leucine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.16</td>\n",
       "      <td>9.1</td>\n",
       "      <td>41.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>lysine</td>\n",
       "      <td>10.8</td>\n",
       "      <td>128.17</td>\n",
       "      <td>5.8</td>\n",
       "      <td>4.2</td>\n",
       "      <td>171.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>methionine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>131.20</td>\n",
       "      <td>2.8</td>\n",
       "      <td>50.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>phenylalanine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147.18</td>\n",
       "      <td>3.9</td>\n",
       "      <td>48.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>proline</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.12</td>\n",
       "      <td>5.1</td>\n",
       "      <td>24.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>serine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87.08</td>\n",
       "      <td>7.4</td>\n",
       "      <td>24.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>threonine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>101.11</td>\n",
       "      <td>6.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>tryptophan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>186.21</td>\n",
       "      <td>1.3</td>\n",
       "      <td>23.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>tyrosine</td>\n",
       "      <td>10.9</td>\n",
       "      <td>163.06</td>\n",
       "      <td>3.3</td>\n",
       "      <td>13.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>valine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.13</td>\n",
       "      <td>6.5</td>\n",
       "      <td>56.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   full_name   pka    mass  occurrence  percent_buried  \\\n",
       "single_letter                                                            \n",
       "A                    alanine   NaN   71.08         7.5            38.0   \n",
       "R                   argenine  12.5  156.10         5.2             0.0   \n",
       "N                 asparagine   NaN  114.04         4.6            10.0   \n",
       "D              aspartic acid   3.9  115.08         5.2            14.5   \n",
       "C                   cysteine   8.3  103.01         1.8            47.0   \n",
       "Q                  glutamine   NaN  128.06         4.1             6.3   \n",
       "E              glutamic acid   4.3  129.12         6.3            20.0   \n",
       "G                    glycine   NaN   57.05         7.1            37.0   \n",
       "H                  histidine   6.0  137.16         2.2            19.0   \n",
       "I                 isoleucine   NaN  113.16         5.5            65.0   \n",
       "L                    leucine   NaN  113.16         9.1            41.0   \n",
       "K                     lysine  10.8  128.17         5.8             4.2   \n",
       "M                 methionine   NaN  131.20         2.8            50.0   \n",
       "F              phenylalanine   NaN  147.18         3.9            48.0   \n",
       "P                    proline   NaN   97.12         5.1            24.0   \n",
       "S                     serine   NaN   87.08         7.4            24.0   \n",
       "T                  threonine   NaN  101.11         6.0            25.0   \n",
       "W                 tryptophan   NaN  186.21         1.3            23.0   \n",
       "Y                   tyrosine  10.9  163.06         3.3            13.0   \n",
       "V                     valine   NaN   99.13         6.5            56.0   \n",
       "\n",
       "               buried_vol    vdw  accessible  hydrophobicity  \n",
       "single_letter                                                 \n",
       "A                    92.0   67.0        67.0             9.0  \n",
       "R                   225.0  148.0       196.0            15.0  \n",
       "N                   135.0   96.0       113.0            16.0  \n",
       "D                   125.0   91.0       106.0            19.0  \n",
       "C                   106.0   86.0       104.0             7.0  \n",
       "Q                   161.0  114.0       144.0            17.0  \n",
       "E                   155.0  109.0       138.0            18.0  \n",
       "G                    66.0   48.0         NaN            11.0  \n",
       "H                   167.0  118.0       151.0            10.0  \n",
       "I                   169.0  124.0       140.0             1.0  \n",
       "L                   168.0  124.0       137.0             3.0  \n",
       "K                   171.0  135.0       167.0            20.0  \n",
       "M                   171.0  124.0       160.0             5.0  \n",
       "F                   203.0  135.0       175.0             2.0  \n",
       "P                   129.0   90.0       105.0            13.0  \n",
       "S                    99.0   73.0        80.0            14.0  \n",
       "T                   122.0   93.0       102.0            12.0  \n",
       "W                   240.0  163.0       217.0             6.0  \n",
       "Y                   203.0  141.0       187.0             8.0  \n",
       "V                   142.0  105.0       117.0             4.0  "
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_props = pd.read_csv(\"../../data/amino_acid_properties.csv\")\n",
    "aa_props.set_index('single_letter', inplace=True)\n",
    "aa_props"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define a function that determines if change in hydrophobicity is above a threshold\n",
    "We arbitrarily set this threshold at 5.0 by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_change_in_hydrophobicity(resi1, resi2, min_diff=5.0):\n",
    "    \"\"\"This function takes string-type amino acid identifiers `resi1` and `resi2`\n",
    "    and compares their hydrophobicities. If the absolute value of the difference\n",
    "    between hydrophobicities is greater than `min_diff`, return boolean True.\n",
    "    Otherwise, return boolean False.\n",
    "    \"\"\"\n",
    "    print(f\"comparing hydrophobicity between {resi1} and {resi2}\")\n",
    "    h1 = aa_props.loc[[resi1]]['hydrophobicity'].item()\n",
    "    h2 = aa_props.loc[[resi2]]['hydrophobicity'].item()\n",
    "    \n",
    "    diff = abs(h1 - h2)\n",
    "    print(f\"the difference is hydrophobicity is {diff}\")\n",
    "    \n",
    "    if diff > min_diff:\n",
    "        return True\n",
    "    else:\n",
    "        return False    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define residues of interest\n",
    "Note: this should be converted to a zero-indexed list for positions in human isoform 2. Most numberings in literature (and on the [UniProt page](https://www.uniprot.org/uniprot/O15393#ptm_processing)) use 1 indexed isoform 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 0 indexed for isoform 2\n",
    "# catalytic_triad = [332, 381, 471]\n",
    "# binding_resi = [259, 260]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1 indexed for isoform 1\n",
    "catalytic_triad = [296, 345, 441]\n",
    "binding_resi = [435, 223, 224]\n",
    "disulfide_bonds = [113, 126, 120, 139, 133,\n",
    "                   148, 172, 231, 185, 241,\n",
    "                   244, 365, 281, 297, 410,\n",
    "                   426, 437, 465]\n",
    "glycosylation_sites = [213, 249]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's convert the index to 0-indexed for isoform 2, simply by adding 36 to each index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[332,\n",
       " 381,\n",
       " 477,\n",
       " 471,\n",
       " 259,\n",
       " 260,\n",
       " 149,\n",
       " 162,\n",
       " 156,\n",
       " 175,\n",
       " 169,\n",
       " 184,\n",
       " 208,\n",
       " 267,\n",
       " 221,\n",
       " 277,\n",
       " 280,\n",
       " 401,\n",
       " 317,\n",
       " 333,\n",
       " 446,\n",
       " 462,\n",
       " 473,\n",
       " 501]"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resi_interest_iso1 = catalytic_triad + binding_resi + disulfide_bonds\n",
    "resi_interest = [idx + 36 for idx in resi_interest_iso1]\n",
    "resi_interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the `compare_resi_prop` function\n",
    "- Arguments\n",
    "    - Two `Seq` instances\n",
    "- Returns\n",
    "    - Pandas dataframe `resi_prop_delta`: integer index at each site, each column is change in residue property for that site\n",
    "    \n",
    "## Define the `get_struc_similarity` function\n",
    "- Arguments\n",
    "    - File path to `resi_prop_delta` CSV\n",
    "- Returns\n",
    "    - Float"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_mutation(resi1, resi2, **columns):\n",
    "    \"\"\"Given two amino acids `resi1` and `resi2`, return a dictionary of\n",
    "    change in amino acid properties between these residues. Use single-letter\n",
    "    amino acid codes. Returns dict.\n",
    "    \"\"\"\n",
    "    assert isinstance(resi1, str)\n",
    "    assert len(resi1) == 1\n",
    "    assert isinstance(resi2, str)\n",
    "    assert len(resi2) == 1\n",
    "    \n",
    "    diff = {\n",
    "        'resi_reference': resi1,\n",
    "        'resi_compare': resi2,\n",
    "    }\n",
    "    \n",
    "    if resi2 == '-':\n",
    "        diff['is_deletion'] = True\n",
    "        columns.update(diff)\n",
    "        return columns\n",
    "    else:\n",
    "        diff['is_deletion'] = False\n",
    "    \n",
    "    dict1 = aa_props.loc[resi1].to_dict()\n",
    "    dict2 = aa_props.loc[resi2].to_dict()\n",
    "    \n",
    "    for k in dict1:\n",
    "        if not isinstance(dict1[k], Number):\n",
    "            continue\n",
    "        diff[f\"{k}_delta\"] = dict2[k] - dict1[k]\n",
    "    \n",
    "    columns.update(diff)\n",
    "    return columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's test this function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'resi_reference': 'Y',\n",
       " 'resi_compare': 'S',\n",
       " 'is_deletion': False,\n",
       " 'pka_delta': nan,\n",
       " 'mass_delta': -75.98,\n",
       " 'occurrence_delta': 4.1000000000000005,\n",
       " 'percent_buried_delta': 11.0,\n",
       " 'buried_vol_delta': -104.0,\n",
       " 'vdw_delta': -68.0,\n",
       " 'accessible_delta': -107.0,\n",
       " 'hydrophobicity_delta': 6.0}"
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_mutation('Y', 'S')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_resi_prop(seq1, seq2, resi_interest=None):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # make sure both of the arguments are SeqRecords\n",
    "    assert isinstance(seq1, SeqRecord.SeqRecord)\n",
    "    assert isinstance(seq2, SeqRecord.SeqRecord)\n",
    "    \n",
    "    # we want to keep track of which amino acid our\n",
    "    # \"cursor\" is on in the for loop\n",
    "    position_counter = 0\n",
    "\n",
    "    # get the entire list of positions in the human sequence as\n",
    "    # integers. We include dashes in this calculation\n",
    "    list_of_positions_including_dashes = range(len(seq1))\n",
    "    \n",
    "    # List of rows (as dicts) that will be converted to a dataframe\n",
    "    rows_list = list()\n",
    "\n",
    "    for position_with_dashes in list_of_positions_including_dashes:\n",
    "        \n",
    "        # get the amino acid at this position (dashes included)\n",
    "        # in both human and cat\n",
    "        resi_in_1 = seq1[position_with_dashes]\n",
    "        resi_in_2 = seq2[position_with_dashes]\n",
    "\n",
    "        # skip this position if it is a '-'\n",
    "        # in the seq1 record\n",
    "        if resi_in_1 == '-':\n",
    "            continue\n",
    "            \n",
    "        # detect if we are at an important amino acid\n",
    "        is_interesting = bool(position_counter in resi_interest)\n",
    "        \n",
    "        # analyze the mutation using a different function\n",
    "        row_as_dict = parse_mutation(resi_in_1, resi_in_2,\n",
    "                                     iso2_zero_idx=position_counter,\n",
    "                                     is_interesting=is_interesting)\n",
    "        rows_list.append(row_as_dict)\n",
    "        \n",
    "        position_counter += 1\n",
    "    \n",
    "    # Convert the list of dicts to a pandas dataframe\n",
    "    df = pd.DataFrame(rows_list)\n",
    "    \n",
    "    # Back-calculate what the position would be in isoform 1\n",
    "    # Recall from the seq sim demo that the offset to account for\n",
    "    # isoform 2 and zero indexing is 36\n",
    "    df['iso1_one_idx'] = df['iso2_zero_idx'] - 36\n",
    "    \n",
    "    # Manually account for negative values\n",
    "    df.loc[df['iso1_one_idx'] <= 1, 'iso1_one_idx'] = np.nan\n",
    "    \n",
    "    # ...but remember that iso2_zero_idx == is actually the first residue...\n",
    "    df.loc[0, 'iso1_one_idx'] = 1\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will test this function below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iso2_zero_idx</th>\n",
       "      <th>is_interesting</th>\n",
       "      <th>resi_reference</th>\n",
       "      <th>resi_compare</th>\n",
       "      <th>is_deletion</th>\n",
       "      <th>pka_delta</th>\n",
       "      <th>mass_delta</th>\n",
       "      <th>occurrence_delta</th>\n",
       "      <th>percent_buried_delta</th>\n",
       "      <th>buried_vol_delta</th>\n",
       "      <th>vdw_delta</th>\n",
       "      <th>accessible_delta</th>\n",
       "      <th>hydrophobicity_delta</th>\n",
       "      <th>iso1_one_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-18.04</td>\n",
       "      <td>6.3</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>P</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-10.04</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-30.0</td>\n",
       "      <td>-17.0</td>\n",
       "      <td>-25.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>P</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>58.98</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-24.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>M</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.12</td>\n",
       "      <td>-4.7</td>\n",
       "      <td>12.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>524</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>488.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>525</td>\n",
       "      <td>False</td>\n",
       "      <td>R</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>489.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>526</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>490.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>527</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>491.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>528</td>\n",
       "      <td>False</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>492.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>529 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     iso2_zero_idx  is_interesting resi_reference resi_compare  is_deletion  \\\n",
       "0                0           False              M            L        False   \n",
       "1                1           False              P            S        False   \n",
       "2                2           False              P            R        False   \n",
       "3                3           False              A            M        False   \n",
       "4                4           False              P            P        False   \n",
       "..             ...             ...            ...          ...          ...   \n",
       "524            524           False              M            -         True   \n",
       "525            525           False              R            -         True   \n",
       "526            526           False              A            -         True   \n",
       "527            527           False              D            -         True   \n",
       "528            528           False              G            -         True   \n",
       "\n",
       "     pka_delta  mass_delta  occurrence_delta  percent_buried_delta  \\\n",
       "0          NaN      -18.04               6.3                  -9.0   \n",
       "1          NaN      -10.04               2.3                   0.0   \n",
       "2          NaN       58.98               0.1                 -24.0   \n",
       "3          NaN       60.12              -4.7                  12.0   \n",
       "4          NaN        0.00               0.0                   0.0   \n",
       "..         ...         ...               ...                   ...   \n",
       "524        NaN         NaN               NaN                   NaN   \n",
       "525        NaN         NaN               NaN                   NaN   \n",
       "526        NaN         NaN               NaN                   NaN   \n",
       "527        NaN         NaN               NaN                   NaN   \n",
       "528        NaN         NaN               NaN                   NaN   \n",
       "\n",
       "     buried_vol_delta  vdw_delta  accessible_delta  hydrophobicity_delta  \\\n",
       "0                -3.0        0.0             -23.0                  -2.0   \n",
       "1               -30.0      -17.0             -25.0                   1.0   \n",
       "2                96.0       58.0              91.0                   2.0   \n",
       "3                79.0       57.0              93.0                  -4.0   \n",
       "4                 0.0        0.0               0.0                   0.0   \n",
       "..                ...        ...               ...                   ...   \n",
       "524               NaN        NaN               NaN                   NaN   \n",
       "525               NaN        NaN               NaN                   NaN   \n",
       "526               NaN        NaN               NaN                   NaN   \n",
       "527               NaN        NaN               NaN                   NaN   \n",
       "528               NaN        NaN               NaN                   NaN   \n",
       "\n",
       "     iso1_one_idx  \n",
       "0             1.0  \n",
       "1             NaN  \n",
       "2             NaN  \n",
       "3             NaN  \n",
       "4             NaN  \n",
       "..            ...  \n",
       "524         488.0  \n",
       "525         489.0  \n",
       "526         490.0  \n",
       "527         491.0  \n",
       "528         492.0  \n",
       "\n",
       "[529 rows x 14 columns]"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resi_prop_delta = compare_resi_prop(human_compact, cat_record, resi_interest=resi_interest)\n",
    "resi_prop_delta #[~resi_prop_delta['iso1_one_idx'].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write this dataframe to the `data` folder:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "resi_prop_delta_Bos_taurus.csv\t\t    resi_prop_delta_Gallus_gallus.csv\r\n",
      "resi_prop_delta_Canis_lupus_familiaris.csv  resi_prop_delta_Homo_sapiens.csv\r\n",
      "resi_prop_delta.csv\t\t\t    resi_prop_delta_Mus_musculus.csv\r\n",
      "resi_prop_delta_Equus_caballus.csv\t    resi_prop_delta_summary.csv\r\n",
      "resi_prop_delta_Felis_catus.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls ../../data/proc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {},
   "outputs": [],
   "source": [
    "resi_prop_delta_fp = \"../../data/proc/resi_prop_delta_test.csv\"\n",
    "resi_prop_delta.to_csv(resi_prop_delta_fp, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Putting it all together\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_species(msa, reference_seq, comparison_species_name, resi_interest, seq_trimming):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    assert isinstance(msa, MultipleSeqAlignment)\n",
    "    assert isinstance(comparison_species_name, str)\n",
    "    assert isinstance(resi_interest, list)\n",
    "    assert isinstance(seq_trimming, slice)\n",
    "\n",
    "    comparison_best_match = get_best_match_by_description(msa, reference_seq, comparison_species_name, seq_trimming)\n",
    "    return compare_resi_prop(reference_seq, comparison_best_match, resi_interest=resi_interest)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can repeat the above comparison between human and cat in one line:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assigning sequence with 2.799922224382656 percent identity as best\n"
     ]
    }
   ],
   "source": [
    "df = compare_species(dom_aln, human_iso2, 'Felis catus', resi_interest, slice(index_nterm, index_cterm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iso2_zero_idx</th>\n",
       "      <th>is_interesting</th>\n",
       "      <th>resi_reference</th>\n",
       "      <th>resi_compare</th>\n",
       "      <th>is_deletion</th>\n",
       "      <th>pka_delta</th>\n",
       "      <th>mass_delta</th>\n",
       "      <th>occurrence_delta</th>\n",
       "      <th>percent_buried_delta</th>\n",
       "      <th>buried_vol_delta</th>\n",
       "      <th>vdw_delta</th>\n",
       "      <th>accessible_delta</th>\n",
       "      <th>hydrophobicity_delta</th>\n",
       "      <th>iso1_one_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>149</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>H</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.3</td>\n",
       "      <td>34.15</td>\n",
       "      <td>0.4</td>\n",
       "      <td>-28.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>113.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>156</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>P</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-5.89</td>\n",
       "      <td>3.3</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>162</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>169</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>133.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>175</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>139.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>184</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>148.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>208</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>267</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>231.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>277</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>241.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>280</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>244.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>317</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>333</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>297.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>401</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>365.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>446</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>410.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>462</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>426.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>473</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>437.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501</th>\n",
       "      <td>501</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>465.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     iso2_zero_idx  is_interesting resi_reference resi_compare  is_deletion  \\\n",
       "149            149            True              C            H        False   \n",
       "156            156            True              C            P        False   \n",
       "162            162            True              C            C        False   \n",
       "169            169            True              C            C        False   \n",
       "175            175            True              C            C        False   \n",
       "184            184            True              C            C        False   \n",
       "208            208            True              C            C        False   \n",
       "221            221            True              C            C        False   \n",
       "267            267            True              C            C        False   \n",
       "277            277            True              C            C        False   \n",
       "280            280            True              C            C        False   \n",
       "317            317            True              C            C        False   \n",
       "333            333            True              C            C        False   \n",
       "401            401            True              C            C        False   \n",
       "446            446            True              C            C        False   \n",
       "462            462            True              C            C        False   \n",
       "473            473            True              C            C        False   \n",
       "501            501            True              C            C        False   \n",
       "\n",
       "     pka_delta  mass_delta  occurrence_delta  percent_buried_delta  \\\n",
       "149       -2.3       34.15               0.4                 -28.0   \n",
       "156        NaN       -5.89               3.3                 -23.0   \n",
       "162        0.0        0.00               0.0                   0.0   \n",
       "169        0.0        0.00               0.0                   0.0   \n",
       "175        0.0        0.00               0.0                   0.0   \n",
       "184        0.0        0.00               0.0                   0.0   \n",
       "208        0.0        0.00               0.0                   0.0   \n",
       "221        0.0        0.00               0.0                   0.0   \n",
       "267        0.0        0.00               0.0                   0.0   \n",
       "277        0.0        0.00               0.0                   0.0   \n",
       "280        0.0        0.00               0.0                   0.0   \n",
       "317        0.0        0.00               0.0                   0.0   \n",
       "333        0.0        0.00               0.0                   0.0   \n",
       "401        0.0        0.00               0.0                   0.0   \n",
       "446        0.0        0.00               0.0                   0.0   \n",
       "462        0.0        0.00               0.0                   0.0   \n",
       "473        0.0        0.00               0.0                   0.0   \n",
       "501        0.0        0.00               0.0                   0.0   \n",
       "\n",
       "     buried_vol_delta  vdw_delta  accessible_delta  hydrophobicity_delta  \\\n",
       "149              61.0       32.0              47.0                   3.0   \n",
       "156              23.0        4.0               1.0                   6.0   \n",
       "162               0.0        0.0               0.0                   0.0   \n",
       "169               0.0        0.0               0.0                   0.0   \n",
       "175               0.0        0.0               0.0                   0.0   \n",
       "184               0.0        0.0               0.0                   0.0   \n",
       "208               0.0        0.0               0.0                   0.0   \n",
       "221               0.0        0.0               0.0                   0.0   \n",
       "267               0.0        0.0               0.0                   0.0   \n",
       "277               0.0        0.0               0.0                   0.0   \n",
       "280               0.0        0.0               0.0                   0.0   \n",
       "317               0.0        0.0               0.0                   0.0   \n",
       "333               0.0        0.0               0.0                   0.0   \n",
       "401               0.0        0.0               0.0                   0.0   \n",
       "446               0.0        0.0               0.0                   0.0   \n",
       "462               0.0        0.0               0.0                   0.0   \n",
       "473               0.0        0.0               0.0                   0.0   \n",
       "501               0.0        0.0               0.0                   0.0   \n",
       "\n",
       "     iso1_one_idx  \n",
       "149         113.0  \n",
       "156         120.0  \n",
       "162         126.0  \n",
       "169         133.0  \n",
       "175         139.0  \n",
       "184         148.0  \n",
       "208         172.0  \n",
       "221         185.0  \n",
       "267         231.0  \n",
       "277         241.0  \n",
       "280         244.0  \n",
       "317         281.0  \n",
       "333         297.0  \n",
       "401         365.0  \n",
       "446         410.0  \n",
       "462         426.0  \n",
       "473         437.0  \n",
       "501         465.0  "
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['iso1_one_idx'].isin(disulfide_bonds)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run for every species that is not human\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing species Homo sapiens to reference 'Homo sapiens'\n",
      "assigning sequence with 100.0 percent identity as best\n",
      "Writing comparison dataframe to ../../data/proc/resi_prop_delta_Homo_sapiens.csv...\n",
      "Comparing species Mus musculus to reference 'Homo sapiens'\n",
      "assigning sequence with 72.77882797731569 percent identity as best\n",
      "Writing comparison dataframe to ../../data/proc/resi_prop_delta_Mus_musculus.csv...\n",
      "Comparing species Canis lupus familiaris to reference 'Homo sapiens'\n",
      "assigning sequence with 26.65406427221172 percent identity as best\n",
      "Writing comparison dataframe to ../../data/proc/resi_prop_delta_Canis_lupus_familiaris.csv...\n",
      "Comparing species Felis catus to reference 'Homo sapiens'\n",
      "assigning sequence with 27.22117202268431 percent identity as best\n",
      "Writing comparison dataframe to ../../data/proc/resi_prop_delta_Felis_catus.csv...\n",
      "Comparing species Bos taurus to reference 'Homo sapiens'\n",
      "assigning sequence with 72.77882797731569 percent identity as best\n",
      "Writing comparison dataframe to ../../data/proc/resi_prop_delta_Bos_taurus.csv...\n",
      "Comparing species Equus caballus to reference 'Homo sapiens'\n",
      "assigning sequence with 68.99810964083176 percent identity as best\n",
      "Writing comparison dataframe to ../../data/proc/resi_prop_delta_Equus_caballus.csv...\n",
      "Comparing species Gallus gallus to reference 'Homo sapiens'\n",
      "assigning sequence with 53.119092627599244 percent identity as best\n",
      "Writing comparison dataframe to ../../data/proc/resi_prop_delta_Gallus_gallus.csv...\n"
     ]
    }
   ],
   "source": [
    "for species_name in domestic_sp_names:\n",
    "    print(f\"Comparing species {species_name} to reference 'Homo sapiens'\")\n",
    "    compare_df = compare_species(dom_aln, human_iso2, species_name,\n",
    "                                 resi_interest, slice(index_nterm, index_cterm))\n",
    "    species_name_no_space = species_name.replace(\" \", \"_\")\n",
    "    compare_fp = f\"../../data/proc/resi_prop_delta_{species_name_no_space}.csv\"\n",
    "    \n",
    "    header = str(\n",
    "        \"# Dataframe containing comparison between human TMPRSS2\\n\" +\n",
    "        \"# isoform 2 and the sequence in aligned EggNOG sequences\\n\" +\n",
    "        f\"# that had highest % identity with the human homolog, for species {species_name}\\n\" +\n",
    "        \"# Author: Ethan Ho\\n\" +\n",
    "        f\"# Date: {dt.now()}\\n\" \n",
    "        \"# From: 20200723_seq_sim.ipynb\\n\" \n",
    "    )\n",
    "    \n",
    "    print(f\"Writing comparison dataframe to {compare_fp}...\")\n",
    "    with open(compare_fp, 'w') as f:\n",
    "        f.write(header)\n",
    "        compare_df.to_csv(f, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Let's take a look at these dataframes...\n",
    "----\n",
    "Sanity checking and such. Also, let's narrow down the data to only residues of interest and write to a `summary` dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Homo sapiens',\n",
       " 'Mus musculus',\n",
       " 'Canis lupus familiaris',\n",
       " 'Felis catus',\n",
       " 'Bos taurus',\n",
       " 'Equus caballus',\n",
       " 'Gallus gallus']"
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "domestic_sp_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [],
   "source": [
    "compare_df_lst = list()\n",
    "for species_name in domestic_sp_names:\n",
    "    species_name_no_space = species_name.replace(\" \", \"_\")\n",
    "    compare_fp = f\"../../data/proc/resi_prop_delta_{species_name_no_space}.csv\"\n",
    "    filtered_df = pd.read_csv(compare_fp, comment='#')\n",
    "    filtered_df = filtered_df[filtered_df.is_interesting]\n",
    "    filtered_df['species'] = species_name\n",
    "    compare_df_lst.append(filtered_df)\n",
    "compare_df_summary = pd.concat(compare_df_lst)\n",
    "compare_df_summary.to_csv(\"../../data/proc/resi_prop_delta_summary.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we can narrow down to residues of interest that changed for each species:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iso2_zero_idx</th>\n",
       "      <th>is_interesting</th>\n",
       "      <th>resi_reference</th>\n",
       "      <th>resi_compare</th>\n",
       "      <th>is_deletion</th>\n",
       "      <th>pka_delta</th>\n",
       "      <th>mass_delta</th>\n",
       "      <th>occurrence_delta</th>\n",
       "      <th>percent_buried_delta</th>\n",
       "      <th>buried_vol_delta</th>\n",
       "      <th>vdw_delta</th>\n",
       "      <th>accessible_delta</th>\n",
       "      <th>hydrophobicity_delta</th>\n",
       "      <th>iso1_one_idx</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>149</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>H</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.3</td>\n",
       "      <td>34.15</td>\n",
       "      <td>0.4</td>\n",
       "      <td>-28.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>156</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>P</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-5.89</td>\n",
       "      <td>3.3</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>259</td>\n",
       "      <td>True</td>\n",
       "      <td>K</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>223.0</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>260</td>\n",
       "      <td>True</td>\n",
       "      <td>K</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-41.09</td>\n",
       "      <td>1.6</td>\n",
       "      <td>19.8</td>\n",
       "      <td>-72.0</td>\n",
       "      <td>-62.0</td>\n",
       "      <td>-87.0</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>149</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>H</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.3</td>\n",
       "      <td>34.15</td>\n",
       "      <td>0.4</td>\n",
       "      <td>-28.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>Felis catus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>156</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>P</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-5.89</td>\n",
       "      <td>3.3</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>Felis catus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>259</td>\n",
       "      <td>True</td>\n",
       "      <td>K</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>223.0</td>\n",
       "      <td>Felis catus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>260</td>\n",
       "      <td>True</td>\n",
       "      <td>K</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-41.09</td>\n",
       "      <td>1.6</td>\n",
       "      <td>19.8</td>\n",
       "      <td>-72.0</td>\n",
       "      <td>-62.0</td>\n",
       "      <td>-87.0</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>Felis catus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>260</td>\n",
       "      <td>True</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>1.7</td>\n",
       "      <td>27.93</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>54.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>Gallus gallus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     iso2_zero_idx  is_interesting resi_reference resi_compare  is_deletion  \\\n",
       "149            149            True              C            H        False   \n",
       "156            156            True              C            P        False   \n",
       "259            259            True              K            -         True   \n",
       "260            260            True              K            S        False   \n",
       "149            149            True              C            H        False   \n",
       "156            156            True              C            P        False   \n",
       "259            259            True              K            -         True   \n",
       "260            260            True              K            S        False   \n",
       "260            260            True              K            R        False   \n",
       "\n",
       "     pka_delta  mass_delta  occurrence_delta  percent_buried_delta  \\\n",
       "149       -2.3       34.15               0.4                 -28.0   \n",
       "156        NaN       -5.89               3.3                 -23.0   \n",
       "259        NaN         NaN               NaN                   NaN   \n",
       "260        NaN      -41.09               1.6                  19.8   \n",
       "149       -2.3       34.15               0.4                 -28.0   \n",
       "156        NaN       -5.89               3.3                 -23.0   \n",
       "259        NaN         NaN               NaN                   NaN   \n",
       "260        NaN      -41.09               1.6                  19.8   \n",
       "260        1.7       27.93              -0.6                  -4.2   \n",
       "\n",
       "     buried_vol_delta  vdw_delta  accessible_delta  hydrophobicity_delta  \\\n",
       "149              61.0       32.0              47.0                   3.0   \n",
       "156              23.0        4.0               1.0                   6.0   \n",
       "259               NaN        NaN               NaN                   NaN   \n",
       "260             -72.0      -62.0             -87.0                  -6.0   \n",
       "149              61.0       32.0              47.0                   3.0   \n",
       "156              23.0        4.0               1.0                   6.0   \n",
       "259               NaN        NaN               NaN                   NaN   \n",
       "260             -72.0      -62.0             -87.0                  -6.0   \n",
       "260              54.0       13.0              29.0                  -5.0   \n",
       "\n",
       "     iso1_one_idx                 species  \n",
       "149         113.0  Canis lupus familiaris  \n",
       "156         120.0  Canis lupus familiaris  \n",
       "259         223.0  Canis lupus familiaris  \n",
       "260         224.0  Canis lupus familiaris  \n",
       "149         113.0             Felis catus  \n",
       "156         120.0             Felis catus  \n",
       "259         223.0             Felis catus  \n",
       "260         224.0             Felis catus  \n",
       "260         224.0           Gallus gallus  "
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_summary_df = compare_df_summary[compare_df_summary.resi_reference != compare_df_summary.resi_compare]\n",
    "mutation_summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing comparison dataframe to ../../data/proc/resi_prop_interesting_mutations.csv...\n"
     ]
    }
   ],
   "source": [
    "mutation_summary_header = str(\n",
    "    \"# Dataframe containing comparison between human TMPRSS2\\n\" +\n",
    "    \"# isoform 2 to other species' homologs.\\n\" +\n",
    "    \"# Filtered to only sequences of interest that are different from the human sequence\\n\" +\n",
    "    \"# Author: Ethan Ho\\n\" +\n",
    "    f\"# Date: {dt.now()}\\n\" \n",
    "    \"# From: 20200723_seq_sim.ipynb\\n\" \n",
    ")\n",
    "\n",
    "mutation_summary_fp = \"../../data/proc/resi_prop_interesting_mutations.csv\"\n",
    "print(f\"Writing comparison dataframe to {mutation_summary_fp}...\")\n",
    "with open(mutation_summary_fp, 'w') as f:\n",
    "    f.write(mutation_summary_header)\n",
    "    mutation_summary_df.to_csv(f, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
